// @inst-iot/bosch-angular-ui-components/esm2015/atoms/modal/modal.module.js
import * as tslib_1 from "tslib";
var ModalModule_1;
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ModalComponent } from './modal.component';
import { ModalDirective } from './modal.directive';
import { ModalService } from './modal.service';
import { IconModule } from '../icon/icon.module';
import { DialogComponent } from './dialog/dialog.component';
import { AlertComponent } from './alert/alert.component';
let ModalModule = ModalModule_1 = class ModalModule {
    static forRoot() {
        return {
            ngModule: ModalModule_1,
            providers: [
                ModalService
            ]
        };
    }
};
ModalModule = ModalModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [
            CommonModule,
            IconModule
        ],
        declarations: [
            ModalComponent,
            ModalDirective,
            DialogComponent,
            AlertComponent
        ],
        exports: [
            ModalComponent,
            ModalDirective,
            DialogComponent,
            AlertComponent
        ],
        entryComponents: [
            ModalComponent
        ]
    })
], ModalModule);
export { ModalModule };
