// @inst-iot/bosch-angular-ui-components/esm2015/molecules/tab-panel/tab-panel-router-outlet.directive.js
import * as tslib_1 from "tslib";
import { ChildrenOutletContexts, RouterOutlet } from '@angular/router';
import { Attribute, Directive } from '@angular/core';
/**
 * Mocks the Router Outlet
 */
let TabPanelRouterOutletDirective = class TabPanelRouterOutletDirective extends RouterOutlet {
    constructor(_parentContexts, name) {
        super(_parentContexts, null, null, name, null);
        this._parentContexts = _parentContexts;
        this.__activatedRoute = null;
        this._name = name;
    }
    ngOnInit() {
        if (!this.__activated) {
            // If the outlet was not instantiated at the time the route got activated we need to populate
            // the outlet when it is initialized (ie inside a NgIf)
            const context = this._parentContexts.getContext(this._name);
            if (context && context.route) {
                if (context.attachRef) {
                    // `attachRef` is populated when there is an existing component to mount
                    this.attach(context.attachRef, context.route);
                }
                else {
                    // otherwise the component defined in the configuration is created
                    this.activateWith(context.route, context.resolver || null);
                }
            }
        }
    }
    get isActivated() {
        return this.__activated;
    }
    get component() {
        if (!this.__activated) {
            throw new Error('Outlet is not activated');
        }
        return null;
    }
    get activatedRoute() {
        if (!this.__activated) {
            throw new Error('Outlet is not activated');
        }
        return this.__activatedRoute;
    }
    get activatedRouteData() {
        if (this.__activatedRoute) {
            return this.__activatedRoute.snapshot.data;
        }
        return {};
    }
    /**
     * Called when the `RouteReuseStrategy` instructs to detach the subtree
     */
    detach() {
        if (!this.__activated) {
            throw new Error('Outlet is not activated');
        }
        this.__activated = null;
        this.__activatedRoute = null;
        return null;
    }
    /**
     * Called when the `RouteReuseStrategy` instructs to re-attach a previously detached subtree
     */
    attach(ref, activatedRoute) {
        this.__activated = true;
        this.__activatedRoute = activatedRoute;
        this.onActivate(activatedRoute);
    }
    deactivate() {
        if (this.__activated) {
            this.__activated = false;
            this.__activatedRoute = null;
        }
    }
    activateWith(activatedRoute, resolver) {
        console.log('activateWith', activatedRoute);
        if (this.isActivated) {
            throw new Error('Cannot activate an already activated outlet');
        }
        this.__activatedRoute = activatedRoute;
        this.__activated = true;
        this.onActivate(activatedRoute);
    }
    onActivate(activatedRoute) {
    }
};
TabPanelRouterOutletDirective = tslib_1.__decorate([
    Directive({
        // tslint:disable-next-line
        selector: 'rb-tab-panel-outlet'
    }),
    tslib_1.__param(1, Attribute('name')),
    tslib_1.__metadata("design:paramtypes", [ChildrenOutletContexts, String])
], TabPanelRouterOutletDirective);
export { TabPanelRouterOutletDirective };
