// @inst-iot/bosch-angular-ui-components/esm2015/molecules/tab-panel/tab-panel-item.directive.js
import * as tslib_1 from "tslib";
import { ContentChild, Directive, Input, TemplateRef } from '@angular/core';
import { TabPanelTitleDirective } from './tab-panel-title.directive';
let tabItemNum = 0;
let TabPanelItemDirective = class TabPanelItemDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.rbTabPanelItemPosition = 'left';
        this.rbTabPanelItemType = 'tab';
        this.rbTabPanelItemAutoClose = true; // dropdown option
        this.routerLink = null;
        this.active = false;
        this.shown = 0;
        this.titleTpl = null;
        this.titleStr = null;
        this.id = 'tabItem' + (tabItemNum++);
    }
    set titleItem(title) {
        if (title && !this.titleTpl) {
            this.titleTpl = title.templateRef;
        }
    }
    set rbTabPanelItemDetails(details) {
        this.details = details;
    }
    set rbTabPanelItemId(id) {
        this.id = id;
    }
    set rbTabPanelItem(title) {
        if (title instanceof TemplateRef) {
            this.titleTpl = title;
        }
        else {
            this.titleStr = title;
        }
    }
    show() {
        this.active = true;
        this.shown++;
    }
    hide() {
        this.active = false;
    }
    get position() {
        return this.rbTabPanelItemPosition;
    }
    get type() {
        return this.rbTabPanelItemType;
    }
    get autoClose() {
        return this.rbTabPanelItemAutoClose;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], TabPanelItemDirective.prototype, "rbTabPanelItemPosition", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], TabPanelItemDirective.prototype, "rbTabPanelItemType", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], TabPanelItemDirective.prototype, "rbTabPanelItemAutoClose", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], TabPanelItemDirective.prototype, "rbTabPanelItemQueryParams", void 0);
tslib_1.__decorate([
    ContentChild(TabPanelTitleDirective, { static: false }),
    tslib_1.__metadata("design:type", TabPanelTitleDirective),
    tslib_1.__metadata("design:paramtypes", [TabPanelTitleDirective])
], TabPanelItemDirective.prototype, "titleItem", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], TabPanelItemDirective.prototype, "rbTabPanelItemDetails", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], TabPanelItemDirective.prototype, "rbTabPanelItemId", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], TabPanelItemDirective.prototype, "rbTabPanelItem", null);
TabPanelItemDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbTabPanelItem]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], TabPanelItemDirective);
export { TabPanelItemDirective };
