// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/form-checkbox/form-checkbox.component.js
import * as tslib_1 from "tslib";
var FormCheckboxComponent_1;
import { Component, ContentChildren, ElementRef, forwardRef, Input, QueryList, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FormValidationMessageDirective } from '../form-validation-message.directive';
import { emptyFunction } from '../forms-util';
let FormCheckboxComponent = FormCheckboxComponent_1 = class FormCheckboxComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.id = 'checkbox.' + Math.random();
        this.label = null;
        this.onChange = emptyFunction;
        this.onTouched = emptyFunction;
    }
    updateValue(value) {
        this.onChange(value);
    }
    isLabelTemplate() {
        return this.label instanceof TemplateRef;
    }
    writeValue(value) {
        this.renderer.setProperty(this.input.nativeElement, 'checked', value);
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.input.nativeElement, 'disabled', isDisabled);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], FormCheckboxComponent.prototype, "required", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], FormCheckboxComponent.prototype, "noValidation", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormCheckboxComponent.prototype, "name", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormCheckboxComponent.prototype, "label", void 0);
tslib_1.__decorate([
    ViewChild('input', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], FormCheckboxComponent.prototype, "input", void 0);
tslib_1.__decorate([
    ContentChildren(FormValidationMessageDirective),
    tslib_1.__metadata("design:type", QueryList)
], FormCheckboxComponent.prototype, "messages", void 0);
FormCheckboxComponent = FormCheckboxComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'rb-form-checkbox',
        template: "<div class=\"input-wrapper\" [class.disabled]=\"input.disabled\">\r\n  <label [for]=\"id + 'input'\">\r\n    <input type=\"checkbox\" class=\"checkbox-control\"\r\n           (blur)=\"onTouched()\" (change)=\"updateValue($event.target.checked)\"\r\n           [id]=\"id + 'input'\" #input>\r\n\r\n    <span class=\"checkbox-icon\">\r\n      <span class=\"check rb-ic rb-ic-checkmark\"></span>\r\n    </span>\r\n\r\n    <span class=\"label-text\">\r\n      {{!isLabelTemplate()?label:''}}\r\n      <ng-container *ngIf=\"isLabelTemplate()\">\r\n        <ng-container *ngTemplateOutlet=\"label\"></ng-container>\r\n      </ng-container>\r\n      <ng-content></ng-content>\r\n    </span>\r\n\r\n  </label>\r\n\r\n</div>\r\n<rb-form-errors [messages]=\"messages\"></rb-form-errors>\r\n",
        providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => FormCheckboxComponent_1), multi: true }]
    }),
    tslib_1.__metadata("design:paramtypes", [Renderer2])
], FormCheckboxComponent);
export { FormCheckboxComponent };
