// @inst-iot/bosch-angular-ui-components/esm5/molecules/accordion/accordion.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { AccordionTitleComponent } from './accordion-title/accordion-title.component';
import { AccordionBodyComponent } from './accordion-body/accordion-body.component';
import { AccordionService } from './accordion.service';
/**
 * <rb-accordion> is the container for
 * <rb-accordion-title> and <rb-accordion-body>
 */
var AccordionComponent = /** @class */ (function () {
    function AccordionComponent(accordion) {
        this.accordion = accordion;
        /**
         * Whether only one can be opened.
         */
        this.openSingle = false;
        this.updateSub = null;
    }
    AccordionComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.updateSub = this.accordion.updateState.subscribe(function (t) { return _this.updateState(t); });
    };
    AccordionComponent.prototype.ngOnDestroy = function () {
        if (this.updateSub) {
            this.updateSub.unsubscribe();
        }
    };
    AccordionComponent.prototype.updateState = function (title) {
        var titles = this.titles.toArray();
        var bodies = this.bodies.toArray();
        var index = titles.indexOf(title);
        if (index !== -1) {
            var component = bodies[index];
            if (title.open) {
                component.open();
            }
            else {
                component.close();
            }
            if (this.openSingle) {
                for (var i = 0; i < titles.length; i++) {
                    var title2 = titles[i];
                    var body = bodies[i];
                    if (i !== index && title2.open) {
                        title2.open = false;
                        body.close();
                    }
                }
            }
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], AccordionComponent.prototype, "openSingle", void 0);
    tslib_1.__decorate([
        ContentChildren(AccordionTitleComponent),
        tslib_1.__metadata("design:type", QueryList)
    ], AccordionComponent.prototype, "titles", void 0);
    tslib_1.__decorate([
        ContentChildren(AccordionBodyComponent),
        tslib_1.__metadata("design:type", QueryList)
    ], AccordionComponent.prototype, "bodies", void 0);
    AccordionComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-accordion',
            template: "<ng-content></ng-content>\r\n",
            providers: [AccordionService]
        }),
        tslib_1.__metadata("design:paramtypes", [AccordionService])
    ], AccordionComponent);
    return AccordionComponent;
}());
export { AccordionComponent };
