// @inst-iot/bosch-angular-ui-components/esm5/atoms/loading-spinner/loading-spinner.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var LoadingSpinnerComponent = /** @class */ (function () {
    function LoadingSpinnerComponent() {
        this.size = 1;
    }
    LoadingSpinnerComponent.prototype.ngOnInit = function () {
    };
    LoadingSpinnerComponent.prototype.getHeight = function () {
        return this.calculate(60);
    };
    LoadingSpinnerComponent.prototype.getWidth = function () {
        return this.calculate(60);
    };
    LoadingSpinnerComponent.prototype.calculate = function (defaultVal) {
        if (this.size > 0 && this.size < 3) {
            return (defaultVal * this.size) + 'px';
        }
        else {
            return defaultVal + 'px';
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LoadingSpinnerComponent.prototype, "size", void 0);
    LoadingSpinnerComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-loading-spinner',
            template: "<div class=\"rb-loader\" [ngStyle]=\"{width: getWidth(), height: getHeight()}\">\n  <div class=\"block1\"></div>\n  <div class=\"block2\"></div>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LoadingSpinnerComponent);
    return LoadingSpinnerComponent;
}());
export { LoadingSpinnerComponent };
