// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/form-errors/form-errors.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, Optional, QueryList } from '@angular/core';
import { NgControl } from '@angular/forms';
var FormErrorsComponent = /** @class */ (function () {
    function FormErrorsComponent(control, cd) {
        this.control = control;
        this.cd = cd;
        this.errors = [];
    }
    FormErrorsComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (this.control) {
            this.sub = this.control.statusChanges.subscribe(function (value) {
                _this.errors = _this.getErrors();
                _this.cd.markForCheck();
            });
        }
    };
    FormErrorsComponent.prototype.ngOnDestroy = function () {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    };
    FormErrorsComponent.prototype.ngDoCheck = function () {
        this.errors = this.getErrors();
    };
    FormErrorsComponent.prototype.getErrors = function () {
        var e_1, _a;
        var errors = [];
        if (this.control && this.control.errors && (this.control.touched || this.control.dirty)) {
            var errorKeys = Object.keys(this.control.errors);
            var _loop_1 = function (key) {
                var message = this_1.messages.find(function (item) { return item.name === key; });
                if (message && (message.onTouched || !message.onTouched && this_1.control.dirty)) {
                    errors.push({
                        name: key,
                        data: this_1.control.errors[key],
                        message: message.templateRef
                    });
                }
            };
            var this_1 = this;
            try {
                for (var errorKeys_1 = tslib_1.__values(errorKeys), errorKeys_1_1 = errorKeys_1.next(); !errorKeys_1_1.done; errorKeys_1_1 = errorKeys_1.next()) {
                    var key = errorKeys_1_1.value;
                    _loop_1(key);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (errorKeys_1_1 && !errorKeys_1_1.done && (_a = errorKeys_1.return)) _a.call(errorKeys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return errors;
    };
    FormErrorsComponent.prototype.getItemName = function (i, item) {
        return item.name;
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", QueryList)
    ], FormErrorsComponent.prototype, "messages", void 0);
    FormErrorsComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-form-errors',
            template: "<div class=\"error-messages\" *ngIf=\"control && control.invalid && errors.length\">\n  <div *ngFor=\"let item of errors; trackBy: getItemName\">\n    <ng-container *ngTemplateOutlet=\"item.message; context: item\"></ng-container>\n  </div>\n</div>\n"
        }),
        tslib_1.__param(0, Optional()),
        tslib_1.__metadata("design:paramtypes", [NgControl,
            ChangeDetectorRef])
    ], FormErrorsComponent);
    return FormErrorsComponent;
}());
export { FormErrorsComponent };
