// @inst-iot/bosch-angular-ui-components/esm5/molecules/loading/error-format/error-format.component.js
import * as tslib_1 from "tslib";
import { Component, Inject, Input, Optional } from '@angular/core';
import { DEFAULT_ERROR_FORMATTER, ERROR_FORMATTER } from '../error-format.model';
var ErrorFormatComponent = /** @class */ (function () {
    function ErrorFormatComponent(formatters) {
        this.opened = false;
        if (!formatters) {
            this.formatters = [DEFAULT_ERROR_FORMATTER];
        }
        else {
            this.formatters = formatters.slice();
            this.formatters.push(DEFAULT_ERROR_FORMATTER);
        }
    }
    ErrorFormatComponent.prototype.ngOnInit = function () {
        this.identifyErrorData(this.error);
    };
    ErrorFormatComponent.prototype.identifyErrorData = function (error) {
        if (this.formatters && this.formatters.length) {
            for (var i = 0; i < this.formatters.length; i++) {
                var formatter = this.formatters[i];
                var result = formatter.transform(error);
                if (result !== null) {
                    this.message = result;
                    if (formatter.getDetails) {
                        this.details = formatter.getDetails(error);
                    }
                    return result;
                }
            }
        }
        this.message = String(error);
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ErrorFormatComponent.prototype, "error", void 0);
    ErrorFormatComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-error-format',
            template: "<div class=\"error-message\">{{message}}</div>\n<a *ngIf=\"details\" href=\"javascript:\" class=\"rb-link rb-details-toggle\" [class.open]=\"opened\" (click)=\"opened=!opened\">Details</a>\n<pre *ngIf=\"opened\" class=\"error-details\">{{details}}</pre>\n",
            styles: [".error-details{word-wrap:break-word;word-break:break-all;white-space:pre-wrap}"]
        }),
        tslib_1.__param(0, Optional()), tslib_1.__param(0, Inject(ERROR_FORMATTER)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], ErrorFormatComponent);
    return ErrorFormatComponent;
}());
export { ErrorFormatComponent };
