// @inst-iot/bosch-angular-ui-components/esm5/molecules/breadcrumbs/breadcrumbs.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BreadcrumbsComponent } from './breadcrumbs.component';
import { BreadcrumbsService } from './breadcrumbs.service';
import { RouterModule } from '@angular/router';
var BreadcrumbsModule = /** @class */ (function () {
    function BreadcrumbsModule() {
    }
    BreadcrumbsModule_1 = BreadcrumbsModule;
    BreadcrumbsModule.forRoot = function () {
        return {
            ngModule: BreadcrumbsModule_1,
            providers: [
                BreadcrumbsService
            ]
        };
    };
    var BreadcrumbsModule_1;
    BreadcrumbsModule = BreadcrumbsModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [
                CommonModule,
                RouterModule
            ],
            declarations: [BreadcrumbsComponent],
            exports: [BreadcrumbsComponent]
        })
    ], BreadcrumbsModule);
    return BreadcrumbsModule;
}());
export { BreadcrumbsModule };
