// @inst-iot/bosch-angular-ui-components/esm5/molecules/breadcrumbs/breadcrumbs.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
import { BehaviorSubject } from 'rxjs';
/**
 * When you define routes, add data: {title: 'myTitle'} to define what is displayed in the breadcrumbs.
 * If the title property is not enough for you to determine what should be displayed, you can
 * just extend this service and provide your own. This is probably necessary when you use some i18n tooling.
 *
 * You need to provide this service to use the rb-breadcrumbs component.
 * e.g. BreadcrumbsModule.forRoot()
 */
var BreadcrumbsService = /** @class */ (function () {
    function BreadcrumbsService(router, activeRoute) {
        this.router = router;
        this.activeRoute = activeRoute;
        this.initAutomatedRouteRecognition();
    }
    BreadcrumbsService.prototype.initAutomatedRouteRecognition = function () {
        var _this = this;
        this.router.events.pipe(filter(function (event) { return event instanceof NavigationEnd; }))
            .subscribe(function () { return _this.refresh(); });
        this.breadcrumbs = new BehaviorSubject(this.identifyRoute());
    };
    BreadcrumbsService.prototype.refresh = function () {
        this.breadcrumbs.next(this.identifyRoute());
    };
    BreadcrumbsService.prototype.identifyRoute = function () {
        var breadcrumbs = [];
        var path = ['/'];
        // Traverse route snapshots to get the breadcrumbs path
        for (var route = this.activeRoute.snapshot; route; route = route.firstChild) {
            if (route.outlet === 'primary') {
                path.push.apply(path, tslib_1.__spread(route.url.map(function (segment) { return segment.path; })));
            }
            var element = this.getBreadCrumbElement(route, path);
            if (element) {
                breadcrumbs.push(element);
            }
        }
        return breadcrumbs;
    };
    BreadcrumbsService.prototype.getBreadCrumbElement = function (route, path) {
        var element = {
            urlSegments: path.slice(0),
            title: '...'
        };
        if (route.outlet === 'primary' && route.data && route.data['title']) {
            element.title = typeof (route.data['title']) === 'function' ? route.data['title'](route) : route.data['title'];
            return element;
        }
        return null;
    };
    BreadcrumbsService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [Router,
            ActivatedRoute])
    ], BreadcrumbsService);
    return BreadcrumbsService;
}());
export { BreadcrumbsService };
