// @inst-iot/bosch-angular-ui-components/esm5/organisms/full-header/navigation/navigation.directive.js
import * as tslib_1 from "tslib";
import { Directive, TemplateRef } from '@angular/core';
var MainNavItemsDirective = /** @class */ (function () {
    function MainNavItemsDirective(templateRef) {
        this.templateRef = templateRef;
    }
    MainNavItemsDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbMainNavItems]'
        }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], MainNavItemsDirective);
    return MainNavItemsDirective;
}());
export { MainNavItemsDirective };
var SubNavItemsDirective = /** @class */ (function () {
    function SubNavItemsDirective(templateRef) {
        this.templateRef = templateRef;
    }
    SubNavItemsDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbSubNavItems]'
        }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], SubNavItemsDirective);
    return SubNavItemsDirective;
}());
export { SubNavItemsDirective };
var MetaNavItemsDirective = /** @class */ (function () {
    function MetaNavItemsDirective(templateRef) {
        this.templateRef = templateRef;
    }
    MetaNavItemsDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbMetaNavItems]'
        }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], MetaNavItemsDirective);
    return MetaNavItemsDirective;
}());
export { MetaNavItemsDirective };
var ActionNavItemsDirective = /** @class */ (function () {
    function ActionNavItemsDirective(templateRef) {
        this.templateRef = templateRef;
    }
    ActionNavItemsDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbActionNavItems]'
        }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], ActionNavItemsDirective);
    return ActionNavItemsDirective;
}());
export { ActionNavItemsDirective };
