// @inst-iot/bosch-angular-ui-components/esm2015/molecules/accordion/accordion-title/accordion-title.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { AccordionService } from '../accordion.service';
let AccordionTitleComponent = class AccordionTitleComponent {
    constructor(accordion) {
        this.accordion = accordion;
        this.open = false;
        this.viewInit = false;
    }
    ngOnChanges(changes) {
        if (changes.open && this.viewInit) {
            this.accordion.updateState.next(this);
        }
    }
    ngAfterViewInit() {
        this.viewInit = true;
        if (this.open) {
            setTimeout(() => {
                this.accordion.updateState.next(this);
            });
        }
    }
    toggle() {
        this.open = !this.open;
        this.accordion.updateState.next(this);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], AccordionTitleComponent.prototype, "open", void 0);
AccordionTitleComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-accordion-title',
        template: "<div class=\"accordion-row\" [class.open]=\"open\" (click)=\"toggle()\">\r\n  <h5>\r\n    <ng-content></ng-content>\r\n  </h5>\r\n  <a href=\"javascript:\" class=\"accordion-toggle\">\r\n    <span class=\"rb-ic\" [class.rb-ic-add]=\"!open\" [class.rb-ic-less-minimize]=\"open\"></span>\r\n  </a>\r\n</div>\r\n"
    }),
    tslib_1.__metadata("design:paramtypes", [AccordionService])
], AccordionTitleComponent);
export { AccordionTitleComponent };
