// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/form-errors/form-errors.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, Optional, QueryList } from '@angular/core';
import { NgControl } from '@angular/forms';
let FormErrorsComponent = class FormErrorsComponent {
    constructor(control, cd) {
        this.control = control;
        this.cd = cd;
        this.errors = [];
    }
    ngOnInit() {
        if (this.control) {
            this.sub = this.control.statusChanges.subscribe(value => {
                this.errors = this.getErrors();
                this.cd.markForCheck();
            });
        }
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
    ngDoCheck() {
        this.errors = this.getErrors();
    }
    getErrors() {
        const errors = [];
        if (this.control && this.control.errors && (this.control.touched || this.control.dirty)) {
            const errorKeys = Object.keys(this.control.errors);
            for (const key of errorKeys) {
                const message = this.messages.find(item => item.name === key);
                if (message && (message.onTouched || !message.onTouched && this.control.dirty)) {
                    errors.push({
                        name: key,
                        data: this.control.errors[key],
                        message: message.templateRef
                    });
                }
            }
        }
        return errors;
    }
    getItemName(i, item) {
        return item.name;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", QueryList)
], FormErrorsComponent.prototype, "messages", void 0);
FormErrorsComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-form-errors',
        template: "<div class=\"error-messages\" *ngIf=\"control && control.invalid && errors.length\">\n  <div *ngFor=\"let item of errors; trackBy: getItemName\">\n    <ng-container *ngTemplateOutlet=\"item.message; context: item\"></ng-container>\n  </div>\n</div>\n"
    }),
    tslib_1.__param(0, Optional()),
    tslib_1.__metadata("design:paramtypes", [NgControl,
        ChangeDetectorRef])
], FormErrorsComponent);
export { FormErrorsComponent };
