// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/validators/form-chips-input-pattern-validator.directive.js
import * as tslib_1 from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
/* tslint:disable:directive-selector */
var FormChipsInputPatternValidatorDirective = /** @class */ (function () {
    function FormChipsInputPatternValidatorDirective() {
    }
    FormChipsInputPatternValidatorDirective_1 = FormChipsInputPatternValidatorDirective;
    FormChipsInputPatternValidatorDirective.prototype.validate = function (c) {
        var _this = this;
        var hasInvalid = false;
        if (c.value && this.pattern) {
            c.value.forEach(function (el) {
                if (!el.match(_this.pattern)) {
                    hasInvalid = true;
                }
            });
            if (hasInvalid) {
                return { 'pattern': true };
            }
            else {
                return null;
            }
        }
        return null;
    };
    var FormChipsInputPatternValidatorDirective_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormChipsInputPatternValidatorDirective.prototype, "pattern", void 0);
    FormChipsInputPatternValidatorDirective = FormChipsInputPatternValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: 'rb-form-chips-input[pattern]',
            providers: [{
                    provide: NG_VALIDATORS,
                    useExisting: forwardRef(function () { return FormChipsInputPatternValidatorDirective_1; }),
                    multi: true
                }]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], FormChipsInputPatternValidatorDirective);
    return FormChipsInputPatternValidatorDirective;
}());
export { FormChipsInputPatternValidatorDirective };
