// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/validators/form-checkbox-required-validator.directive.js
import * as tslib_1 from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
/* tslint:disable:directive-selector */
var FormCheckboxRequiredValidatorDirective = /** @class */ (function () {
    function FormCheckboxRequiredValidatorDirective() {
        this.statesMode = 'check';
    }
    FormCheckboxRequiredValidatorDirective_1 = FormCheckboxRequiredValidatorDirective;
    Object.defineProperty(FormCheckboxRequiredValidatorDirective.prototype, "required", {
        get: function () {
            return this._required;
        },
        set: function (value) {
            this._required = value != null && value !== false && "" + value !== 'false';
            if (this._onChange) {
                this._onChange();
            }
        },
        enumerable: true,
        configurable: true
    });
    FormCheckboxRequiredValidatorDirective.prototype.validate = function (c) {
        if (this.required) {
            return this.hasValue(c) ? null : { 'required': true };
        }
        return null;
    };
    FormCheckboxRequiredValidatorDirective.prototype.hasValue = function (c) {
        if (typeof (c.value) === 'boolean') {
            return c.value === true;
        }
        if (typeof (c.value) === 'object' && c.value) {
            var values = Object.keys(c.value).map(function (key) { return c.value[key]; });
            if (this.statesMode === 'optCheck') {
                return values.some(function (v) { return v !== null; });
            }
            if (this.statesMode === 'check') {
                return values.some(function (v) { return v !== false; });
            }
        }
        return true;
    };
    FormCheckboxRequiredValidatorDirective.prototype.registerOnValidatorChange = function (fn) {
        this._onChange = fn;
    };
    var FormCheckboxRequiredValidatorDirective_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], FormCheckboxRequiredValidatorDirective.prototype, "statesMode", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], FormCheckboxRequiredValidatorDirective.prototype, "required", null);
    FormCheckboxRequiredValidatorDirective = FormCheckboxRequiredValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: 'rb-form-checkbox[required],rb-form-multi-select[required]',
            providers: [{
                    provide: NG_VALIDATORS,
                    useExisting: forwardRef(function () { return FormCheckboxRequiredValidatorDirective_1; }),
                    multi: true
                }]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], FormCheckboxRequiredValidatorDirective);
    return FormCheckboxRequiredValidatorDirective;
}());
export { FormCheckboxRequiredValidatorDirective };
