// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/validators/form-number-validator.directive.js
import * as tslib_1 from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
var FormNumberValidatorDirective = /** @class */ (function () {
    function FormNumberValidatorDirective() {
    }
    FormNumberValidatorDirective_1 = FormNumberValidatorDirective;
    FormNumberValidatorDirective.prototype.validate = function (control) {
        if (control.value === '') {
            return null;
        }
        var value = typeof (control.value) === 'number' ? control.value : parseFloat(control.value);
        var errors = {};
        if (this.rbMin !== undefined && value < this.rbMin) {
            errors.rbMin = this.rbMin;
        }
        if (this.rbMax !== undefined && value > this.rbMax) {
            errors.rbMax = this.rbMax;
        }
        return Object.keys(errors).length ? errors : null;
    };
    var FormNumberValidatorDirective_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], FormNumberValidatorDirective.prototype, "rbMin", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], FormNumberValidatorDirective.prototype, "rbMax", void 0);
    FormNumberValidatorDirective = FormNumberValidatorDirective_1 = tslib_1.__decorate([
        Directive({
            selector: '[rbMin],[rbMax]',
            providers: [{
                    provide: NG_VALIDATORS,
                    useExisting: forwardRef(function () { return FormNumberValidatorDirective_1; }),
                    multi: true
                }]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], FormNumberValidatorDirective);
    return FormNumberValidatorDirective;
}());
export { FormNumberValidatorDirective };
