// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/form-radio/form-radio.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, ElementRef, forwardRef, Input, QueryList, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { FormValidationMessageDirective } from '../form-validation-message.directive';
import { emptyFunction } from '../forms-util';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var FormRadioComponent = /** @class */ (function () {
    function FormRadioComponent(renderer) {
        this.renderer = renderer;
        this.id = 'radio.' + Math.random();
        this.label = null;
        this.onChange = emptyFunction;
        this.onTouched = emptyFunction;
    }
    FormRadioComponent_1 = FormRadioComponent;
    FormRadioComponent.prototype.updateValue = function (isChecked) {
        if (isChecked) {
            this.onChange(this.value);
        }
    };
    FormRadioComponent.prototype.isLabelTemplate = function () {
        return this.label instanceof TemplateRef;
    };
    FormRadioComponent.prototype.writeValue = function (value) {
        if (value === this.value) {
            this.renderer.setProperty(this.input.nativeElement, 'checked', true);
        }
    };
    FormRadioComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    FormRadioComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    FormRadioComponent.prototype.setDisabledState = function (isDisabled) {
        this.renderer.setProperty(this.input.nativeElement, 'disabled', isDisabled);
    };
    FormRadioComponent.prototype.isReadonly = function () {
        return !!(this.readonly === '' || this.readonly);
    };
    var FormRadioComponent_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], FormRadioComponent.prototype, "required", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], FormRadioComponent.prototype, "noValidation", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], FormRadioComponent.prototype, "name", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormRadioComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        ViewChild('input', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], FormRadioComponent.prototype, "input", void 0);
    tslib_1.__decorate([
        ContentChildren(FormValidationMessageDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], FormRadioComponent.prototype, "messages", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormRadioComponent.prototype, "value", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormRadioComponent.prototype, "readonly", void 0);
    FormRadioComponent = FormRadioComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'rb-form-radio',
            template: "<div class=\"input-wrapper\">\r\n  <input type=\"radio\" [id]=\"id + 'input'\" [name]=\"name\" [readonly]=\"readonly\"\r\n         (click)=\"updateValue($event.target.checked)\" #input>\r\n  <label [for]=\"id + 'input'\">  {{!isLabelTemplate() ? label : ''}}</label>\r\n</div>\r\n\r\n<rb-form-errors [messages]=\"messages\"></rb-form-errors>\r\n",
            providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(function () { return FormRadioComponent_1; }), multi: true }]
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2])
    ], FormRadioComponent);
    return FormRadioComponent;
}());
export { FormRadioComponent };
