// @inst-iot/bosch-angular-ui-components/esm5/atoms/details-toggle/details-toggle.directive.js
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, HostListener, Input, Output, ViewContainerRef } from '@angular/core';
/**
 * Can be toggled to provide a state
 * Usage with: rbDetailsToggle #triggerDetails="rbDetailsToggle"
 * Then check triggerDetails.open for the current state
 */
var DetailsToggleDirective = /** @class */ (function () {
    function DetailsToggleDirective(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this.isOpenChange = new EventEmitter();
        this.open = false;
    }
    Object.defineProperty(DetailsToggleDirective.prototype, "isOpen", {
        set: function (state) {
            if (state) {
                this.openDetails();
            }
            else {
                this.closeDetails();
            }
        },
        enumerable: true,
        configurable: true
    });
    DetailsToggleDirective.prototype.onClick = function (e) {
        e.preventDefault();
        if (this.open) {
            this.closeDetails();
            this.isOpenChange.next(this.open);
        }
        else {
            this.openDetails();
            this.isOpenChange.next(this.open);
        }
    };
    DetailsToggleDirective.prototype.openDetails = function () {
        this.open = true;
        this.viewContainerRef.element.nativeElement.classList.add('open');
    };
    DetailsToggleDirective.prototype.closeDetails = function () {
        this.open = false;
        this.viewContainerRef.element.nativeElement.classList.remove('open');
    };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], DetailsToggleDirective.prototype, "isOpenChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], DetailsToggleDirective.prototype, "isOpen", null);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [MouseEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], DetailsToggleDirective.prototype, "onClick", null);
    DetailsToggleDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbDetailsToggle]',
            exportAs: 'rbDetailsToggle'
        }),
        tslib_1.__metadata("design:paramtypes", [ViewContainerRef])
    ], DetailsToggleDirective);
    return DetailsToggleDirective;
}());
export { DetailsToggleDirective };
