// @inst-iot/bosch-angular-ui-components/esm5/atoms/progress-indicator/progress-indicator.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var ProgressIndicatorComponent = /** @class */ (function () {
    function ProgressIndicatorComponent() {
        this.total = Infinity;
        this.current = Infinity;
    }
    ProgressIndicatorComponent.prototype.ngOnInit = function () {
    };
    Object.defineProperty(ProgressIndicatorComponent.prototype, "percent", {
        get: function () {
            return this.total > 0 && this.total !== Infinity ? this.current / this.total * 100 : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressIndicatorComponent.prototype, "isDeterminate", {
        get: function () {
            return this.total !== Infinity;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ProgressIndicatorComponent.prototype, "total", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ProgressIndicatorComponent.prototype, "current", void 0);
    ProgressIndicatorComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-progress-indicator',
            template: "<div class=\"progress-container\" [class.noanim]=\"isDeterminate\">\n  <div class=\"progress-current\" [style.width]=\"percent + '%'\">\n    <div class=\"progress-gradient\"></div>\n  </div>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], ProgressIndicatorComponent);
    return ProgressIndicatorComponent;
}());
export { ProgressIndicatorComponent };
