// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-date-fields/date-range-picker/date-range-picker.model.js
export var defaultTimeInputOptions = {
    units: [
        { value: 1, unit: 'ms', label: 'Milliseconds' },
        { value: 1000, unit: 's', label: 'Seconds' },
        { value: 1000 * 60, unit: 'm', label: 'Minutes' },
        { value: 1000 * 60 * 60, unit: 'h', label: 'Hours' },
        { value: 1000 * 60 * 60 * 24, unit: 'd', label: 'Days' },
        { value: 1000 * 60 * 60 * 24 * 7, unit: 'w', label: 'Weeks' },
    ]
};
export function isRelative(values) {
    return Array.isArray(values) && typeof (values[0]) === 'number' && typeof (values[1]) === 'number';
}
export function isAbsolute(values) {
    return Array.isArray(values) && typeof (values[0]) === 'string' && typeof (values[1]) === 'string';
}
export function getRelativeInfo(value, units) {
    var operator = '-';
    var count = 0;
    var unit = 'ms';
    if (typeof (value) === 'string' && value.match(/^([+-])(\d+)/)) {
        operator = RegExp.$1;
        count = parseInt(RegExp.$2, 10);
    }
    if (typeof (value) === 'number') {
        operator = value < 0 ? '-' : '+';
        count = Math.abs(value);
    }
    if (count > 0) {
        var divisor_1 = 1;
        units.forEach(function (u) {
            if (count % u.value === 0) {
                unit = u.unit;
                divisor_1 = u.value;
            }
        });
        count = count / divisor_1;
    }
    var label = units.find(function (u) { return u.unit === unit; }).label;
    return {
        operator: operator,
        count: count,
        unit: unit,
        label: label,
        displayTextShort: "" + operator + count + unit,
        displayText: "" + operator + count + " " + label
    };
}
