// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-date-fields/form-date-input/form-date-input.component.js
import * as tslib_1 from "tslib";
var FormDateInputComponent_1;
import { Component, ContentChildren, ElementRef, forwardRef, Inject, InjectionToken, Input, Optional, QueryList, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FormValidationMessageDirective } from '../../form-fields/form-validation-message.directive';
import { emptyFunction } from '../../form-fields/forms-util';
import * as flatpickrImport from 'flatpickr';
const flatpickrFunc = flatpickrImport; // workaround for rollup and tests
export const FLATPICKR_DEFAULT_OPTIONS = new InjectionToken('flatpickrDefaultOptions');
const overridableOptions = {
    time_24hr: true
};
/**
 * Displays a input for flatpickr
 * The value, is a ISO Date String or an array of ISO Date Strings.
 */
let FormDateInputComponent = FormDateInputComponent_1 = class FormDateInputComponent {
    constructor(renderer, elementRef, defaultOptions) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.defaultOptions = defaultOptions;
        this.label = null;
        this.type = 'text';
        this.id = 'input.' + Math.random();
        /**
         * Options for Flatpickr
         * @see https://flatpickr.js.org/options/
         */
        this.options = {};
        /**
         * Given, when this is the start date.
         * Start input is responsible for options
         */
        this.rangeEnd = null;
        this.onChange = emptyFunction;
        this.onTouched = emptyFunction;
    }
    ngAfterViewInit() {
        if (this.defaultOptions) {
            this.options = Object.assign({}, overridableOptions, this.defaultOptions, this.options);
        }
        else {
            this.options = Object.assign({}, overridableOptions, this.options);
        }
        this.options.onValueUpdate = (selectedDates, dateString) => {
            this.updateValue(selectedDates);
        };
        if (this.rangeEnd) {
            this.rangeEnd.options = Object.assign({}, this.options);
        }
        this.picker = flatpickrFunc(this.input.nativeElement, this.options);
    }
    ngOnChanges(changes) {
        if (changes.options && this.picker && changes.options.currentValue) {
            Object.keys(changes.options.currentValue).forEach(opt => {
                this.picker.set(opt, changes.options.currentValue[opt]);
            });
        }
    }
    ngOnDestroy() {
        if (this.picker) {
            this.picker.destroy();
        }
    }
    isReadonly() {
        return this.readonly !== undefined;
    }
    updateValue(value) {
        if (this.picker.config.mode === 'single' && value && value.length) {
            value = value[0].toISOString();
        }
        if (['range', 'multiple'].indexOf(this.picker.config.mode) !== -1 && value && value.length) {
            value = value.map(v => v.toISOString());
        }
        this.checkValue(value);
        this.onChange(value);
        this.updateEndRange();
    }
    updateEndRange() {
        if (this.rangeEnd && this.picker.selectedDates[0]) {
            const start = this.picker.selectedDates[0];
            const end = this.rangeEnd.picker.selectedDates[0];
            this.rangeEnd.picker.set('minDate', this.picker.selectedDates[0]);
            if (start && end && end.getTime() < start.getTime()) {
                this.rangeEnd.writeValue(start.toISOString());
            }
        }
    }
    checkValue(value) {
        if (typeof (value) === 'string' && value.length > 0 || value && value.length) {
            this.renderer.addClass(this.elementRef.nativeElement, 'not-empty');
        }
        else {
            this.renderer.removeClass(this.elementRef.nativeElement, 'not-empty');
        }
    }
    isLabelTemplate() {
        return this.label instanceof TemplateRef;
    }
    writeValue(value) {
        if (value === null || value === undefined) {
            value = '';
        }
        this.checkValue(value);
        if (this.picker && this.picker.config) {
            this.picker.setDate(value);
        }
        this.updateEndRange();
        // this.renderer.setProperty(this.input.nativeElement, 'value', value);
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.input.nativeElement, 'disabled', isDisabled);
    }
    toggle() {
        this.picker.toggle();
    }
    close() {
        this.picker.close();
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormDateInputComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormDateInputComponent.prototype, "name", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormDateInputComponent.prototype, "type", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormDateInputComponent.prototype, "readonly", void 0);
tslib_1.__decorate([
    ContentChildren(FormValidationMessageDirective),
    tslib_1.__metadata("design:type", QueryList)
], FormDateInputComponent.prototype, "messages", void 0);
tslib_1.__decorate([
    ViewChild('input', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], FormDateInputComponent.prototype, "input", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormDateInputComponent.prototype, "options", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", FormDateInputComponent)
], FormDateInputComponent.prototype, "rangeEnd", void 0);
FormDateInputComponent = FormDateInputComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'rb-form-date-input',
        template: "<div class=\"input-wrapper\" [class.no-label]=\"!label\">\r\n\r\n  <input [type]=\"type\" [id]=\"id + 'input'\"\r\n         class=\"input with-icon\"\r\n         (blur)=\"onTouched()\"\r\n         [readonly]=\"isReadonly()\"\r\n         #input>\r\n  <span class=\"input-icon rb-ic rb-ic-calendar\" (click)=\"toggle()\"></span>\r\n  <label class=\"label\" [for]=\"id + 'input'\">\r\n    {{!isLabelTemplate()?label:''}}\r\n    <ng-container *ngIf=\"isLabelTemplate()\">\r\n      <ng-container *ngTemplateOutlet=\"label\"></ng-container>\r\n    </ng-container>\r\n  </label>\r\n  <span class=\"input-background\"></span>\r\n</div>\r\n\r\n<rb-form-errors [messages]=\"messages\"></rb-form-errors>\r\n",
        providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => FormDateInputComponent_1), multi: true }],
        exportAs: 'dateInput'
    }),
    tslib_1.__param(2, Optional()), tslib_1.__param(2, Inject(FLATPICKR_DEFAULT_OPTIONS)),
    tslib_1.__metadata("design:paramtypes", [Renderer2,
        ElementRef, Object])
], FormDateInputComponent);
export { FormDateInputComponent };
