// @inst-iot/bosch-angular-ui-components/esm2015/organisms/full-header/navigation/navigation.directive.js
import * as tslib_1 from "tslib";
import { Directive, TemplateRef } from '@angular/core';
let MainNavItemsDirective = class MainNavItemsDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
MainNavItemsDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbMainNavItems]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], MainNavItemsDirective);
export { MainNavItemsDirective };
let SubNavItemsDirective = class SubNavItemsDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
SubNavItemsDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbSubNavItems]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], SubNavItemsDirective);
export { SubNavItemsDirective };
let MetaNavItemsDirective = class MetaNavItemsDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
MetaNavItemsDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbMetaNavItems]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], MetaNavItemsDirective);
export { MetaNavItemsDirective };
let ActionNavItemsDirective = class ActionNavItemsDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
ActionNavItemsDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbActionNavItems]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], ActionNavItemsDirective);
export { ActionNavItemsDirective };
