// @inst-iot/bosch-angular-ui-components/esm2015/molecules/loading/error-format.model.js
import { InjectionToken } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
export const ERROR_FORMATTER = new InjectionToken('rbErrorFormatter');
export class DetailedError {
    constructor(message, details) {
        this.message = message;
        this.details = details;
        this.name = 'DetailedError';
    }
    toString() {
        return this.message;
    }
}
export class DefaultErrorFormatter {
    getDetails(error) {
        if (error instanceof DetailedError) {
            if (typeof (error.details) === 'object') {
                return JSON.stringify(error.details, null, 2);
            }
            return error.details;
        }
        if (error instanceof HttpErrorResponse) {
            if (error.error) {
                if (typeof (error.error) === 'object') {
                    return JSON.stringify(error.error, null, 2);
                }
                return String(error.error);
            }
        }
        return null;
    }
    transform(error) {
        if (error instanceof HttpErrorResponse) {
            let text = 'Error: ';
            if (error.message) {
                text += error.message;
            }
            return text;
        }
        if (error instanceof Error || error instanceof DetailedError) {
            return String(error);
        }
        return null;
    }
}
export const DEFAULT_ERROR_FORMATTER = new DefaultErrorFormatter();
