// @inst-iot/bosch-angular-ui-components/esm2015/molecules/loading/loading-utils.js
import { Subject } from 'rxjs';
import { takeUntil, tap } from 'rxjs/operators';
export class LoadingEntity {
    constructor(loadingStateAfterFirstSuccess = true) {
        this.loadingStateAfterFirstSuccess = loadingStateAfterFirstSuccess;
        this.runs = 0;
        this.loading = false;
        this.error = null;
        this.result = null;
        this.stopSubject = new Subject();
    }
    /**
     * Tracks the status of an observable
     */
    run(observable) {
        this.runs++;
        if (this.loadingStateAfterFirstSuccess && this.result !== null || this.result === null || this.error) {
            this.loading = true;
            this.result = null;
        }
        this.complete();
        this.error = null;
        return observable.pipe(tap(result => {
            this.loading = false;
            this.result = result;
        }, error => {
            this.loading = false;
            this.error = error;
        }, () => {
            this.loading = false;
        }), takeUntil(this.stopSubject));
    }
    /**
     * This can complete and unsubscribe all observers
     */
    complete() {
        this.stopSubject.next();
    }
    reset() {
        this.complete();
        this.error = null;
        this.loading = false;
        this.result = null;
    }
}
