// @inst-iot/bosch-angular-ui-components/esm2015/molecules/loading/loading.directive.js
import * as tslib_1 from "tslib";
import { ComponentFactoryResolver, Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { LoadingEntity } from './loading-utils';
import { LoadingStatusComponent } from './loading-status/loading-status.component';
let LoadingDirective = class LoadingDirective {
    constructor(templateRef, viewContainer, componentFactory) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.componentFactory = componentFactory;
        this.loadingRef = null;
        this.viewRef = null;
        this.rbLoadingSize = 1;
    }
    ngDoCheck() {
        if (this.loadingEntity) {
            this.updateView(this.loadingEntity.loading || this.loadingEntity.error);
        }
    }
    set rbLoading(state) {
        this.factory = this.componentFactory.resolveComponentFactory(LoadingStatusComponent);
        this.updateState(state);
    }
    updateState(state) {
        if (state instanceof LoadingEntity) {
            this.loadingEntity = state;
            state = !!(this.loadingEntity.loading || this.loadingEntity.error);
        }
        this.updateView(state);
    }
    updateView(state) {
        if (state) {
            if (!this.loadingRef) {
                this.viewContainer.clear();
                this.viewRef = null;
                this.loadingRef = this.viewContainer.createComponent(this.factory);
                this.loadingRef.instance.loadingEntity = this.loadingEntity;
                this.loadingRef.instance.size = this.rbLoadingSize;
            }
        }
        else {
            if (!this.viewRef) {
                this.viewContainer.clear();
                this.loadingRef = null;
                this.viewRef = this.viewContainer.createEmbeddedView(this.templateRef);
            }
        }
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], LoadingDirective.prototype, "rbLoadingSize", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], LoadingDirective.prototype, "rbLoading", null);
LoadingDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbLoading]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef,
        ViewContainerRef,
        ComponentFactoryResolver])
], LoadingDirective);
export { LoadingDirective };
