// @inst-iot/bosch-angular-ui-components/esm2015/molecules/loading/loading-link.directive.js
import * as tslib_1 from "tslib";
import { ContentChildren, Directive, ElementRef, Input, QueryList, Renderer2 } from '@angular/core';
import { NavigationEnd, NavigationStart, Router, RouterLinkWithHref } from '@angular/router';
let LoadingLinkDirective = class LoadingLinkDirective {
    constructor(router, element, renderer) {
        this.router = router;
        this.element = element;
        this.renderer = renderer;
        this.loading = false;
        this.className = 'loading';
        this.subscription = router.events.subscribe(s => {
            if (s instanceof NavigationStart && this.isRelevantUrl(s.url)) {
                this.loading = true;
                this.update();
            }
            if (s instanceof NavigationEnd) {
                this.loading = false;
                this.update();
            }
        });
    }
    set rbLoadingLink(className) {
        if (className) {
            this.className = className;
        }
    }
    ngAfterContentInit() {
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    update() {
        if (!this.element || !this.element.nativeElement) {
            return;
        }
        if (this.loading) {
            this.renderer.addClass(this.element.nativeElement, this.className);
        }
        else {
            this.renderer.removeClass(this.element.nativeElement, this.className);
        }
    }
    isRelevantUrl(url) {
        return this.linksWithHrefs.some(link => url === link.urlTree.toString());
    }
};
tslib_1.__decorate([
    ContentChildren(RouterLinkWithHref, { descendants: true }),
    tslib_1.__metadata("design:type", QueryList)
], LoadingLinkDirective.prototype, "linksWithHrefs", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LoadingLinkDirective.prototype, "rbLoadingLink", null);
LoadingLinkDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbLoadingLink]'
    }),
    tslib_1.__metadata("design:paramtypes", [Router, ElementRef, Renderer2])
], LoadingLinkDirective);
export { LoadingLinkDirective };
