// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/validators/form-checkbox-required-validator.directive.js
import * as tslib_1 from "tslib";
var FormCheckboxRequiredValidatorDirective_1;
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
/* tslint:disable:directive-selector */
let FormCheckboxRequiredValidatorDirective = FormCheckboxRequiredValidatorDirective_1 = class FormCheckboxRequiredValidatorDirective {
    constructor() {
        this.statesMode = 'check';
    }
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value != null && value !== false && `${value}` !== 'false';
        if (this._onChange) {
            this._onChange();
        }
    }
    validate(c) {
        if (this.required) {
            return this.hasValue(c) ? null : { 'required': true };
        }
        return null;
    }
    hasValue(c) {
        if (typeof (c.value) === 'boolean') {
            return c.value === true;
        }
        if (typeof (c.value) === 'object' && c.value) {
            const values = Object.keys(c.value).map(key => c.value[key]);
            if (this.statesMode === 'optCheck') {
                return values.some(v => v !== null);
            }
            if (this.statesMode === 'check') {
                return values.some(v => v !== false);
            }
        }
        return true;
    }
    registerOnValidatorChange(fn) {
        this._onChange = fn;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormCheckboxRequiredValidatorDirective.prototype, "statesMode", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], FormCheckboxRequiredValidatorDirective.prototype, "required", null);
FormCheckboxRequiredValidatorDirective = FormCheckboxRequiredValidatorDirective_1 = tslib_1.__decorate([
    Directive({
        selector: 'rb-form-checkbox[required],rb-form-multi-select[required]',
        providers: [{
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => FormCheckboxRequiredValidatorDirective_1),
                multi: true
            }]
    }),
    tslib_1.__metadata("design:paramtypes", [])
], FormCheckboxRequiredValidatorDirective);
export { FormCheckboxRequiredValidatorDirective };
