// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/validators/form-file-validator.directive.js
import * as tslib_1 from "tslib";
var FormFileValidatorDirective_1;
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
/* tslint:disable:directive-selector */
let FormFileValidatorDirective = FormFileValidatorDirective_1 = class FormFileValidatorDirective {
    constructor() {
    }
    ngOnChanges(changes) {
        if (this.onChange && (changes.min || changes.max || changes.maxSize)) {
            this.onChange();
        }
    }
    validate(c) {
        const errors = {};
        const minNumber = this.min ? parseInt(this.min, 10) : 0;
        const maxNumber = this.max ? parseInt(this.max, 10) : 0;
        const maxSize = this.maxSize ? parseInt(this.maxSize, 10) : 0;
        if (minNumber && (!c.value || Array.isArray(c.value) && c.value.length < minNumber)) {
            errors.min = true;
        }
        if (maxNumber && (!c.value || Array.isArray(c.value) && c.value.length > maxNumber)) {
            errors.max = true;
        }
        if (maxSize && Array.isArray(c.value)) {
            c.value.some((file) => {
                if (file.size > maxSize) {
                    errors.maxSize = file.name;
                    return true;
                }
                return false;
            });
        }
        return Object.keys(errors).length ? errors : null;
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormFileValidatorDirective.prototype, "min", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormFileValidatorDirective.prototype, "max", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], FormFileValidatorDirective.prototype, "maxSize", void 0);
FormFileValidatorDirective = FormFileValidatorDirective_1 = tslib_1.__decorate([
    Directive({
        selector: 'rb-form-file[min],rb-form-file[max],rb-form-file[min],rb-form-file[maxSize]',
        providers: [{
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => FormFileValidatorDirective_1),
                multi: true
            }]
    }),
    tslib_1.__metadata("design:paramtypes", [])
], FormFileValidatorDirective);
export { FormFileValidatorDirective };
