// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/validators/form-chips-input-pattern-validator.directive.js
import * as tslib_1 from "tslib";
var FormChipsInputPatternValidatorDirective_1;
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
/* tslint:disable:directive-selector */
let FormChipsInputPatternValidatorDirective = FormChipsInputPatternValidatorDirective_1 = class FormChipsInputPatternValidatorDirective {
    constructor() {
    }
    validate(c) {
        let hasInvalid = false;
        if (c.value && this.pattern) {
            c.value.forEach(el => {
                if (!el.match(this.pattern)) {
                    hasInvalid = true;
                }
            });
            if (hasInvalid) {
                return { 'pattern': true };
            }
            else {
                return null;
            }
        }
        return null;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormChipsInputPatternValidatorDirective.prototype, "pattern", void 0);
FormChipsInputPatternValidatorDirective = FormChipsInputPatternValidatorDirective_1 = tslib_1.__decorate([
    Directive({
        selector: 'rb-form-chips-input[pattern]',
        providers: [{
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => FormChipsInputPatternValidatorDirective_1),
                multi: true
            }]
    }),
    tslib_1.__metadata("design:paramtypes", [])
], FormChipsInputPatternValidatorDirective);
export { FormChipsInputPatternValidatorDirective };
