// @inst-iot/bosch-angular-ui-components/esm2015/atoms/progress-indicator/progress-indicator.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
let ProgressIndicatorComponent = class ProgressIndicatorComponent {
    constructor() {
        this.total = Infinity;
        this.current = Infinity;
    }
    ngOnInit() {
    }
    get percent() {
        return this.total > 0 && this.total !== Infinity ? this.current / this.total * 100 : 100;
    }
    get isDeterminate() {
        return this.total !== Infinity;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], ProgressIndicatorComponent.prototype, "total", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], ProgressIndicatorComponent.prototype, "current", void 0);
ProgressIndicatorComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-progress-indicator',
        template: "<div class=\"progress-container\" [class.noanim]=\"isDeterminate\">\n  <div class=\"progress-current\" [style.width]=\"percent + '%'\">\n    <div class=\"progress-gradient\"></div>\n  </div>\n</div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [])
], ProgressIndicatorComponent);
export { ProgressIndicatorComponent };
