// @inst-iot/bosch-angular-ui-components/esm5/atoms/popover/popover.directive.js
import * as tslib_1 from "tslib";
import { ComponentFactoryResolver, Directive, HostListener, Input, ViewContainerRef } from '@angular/core';
import { PopoverComponent } from './popover.component';
var PopoverDirective = /** @class */ (function () {
    function PopoverDirective(factoryResolver, viewContainerRef) {
        this.factoryResolver = factoryResolver;
        this.viewContainerRef = viewContainerRef;
        this.position = 'bottom';
        this.componentRef = null;
        this.closeSub = null;
    }
    PopoverDirective.prototype.ngOnDestroy = function () {
        this.closePopOver();
    };
    Object.defineProperty(PopoverDirective.prototype, "template", {
        set: function (tpl) {
            this.content = tpl;
        },
        enumerable: true,
        configurable: true
    });
    PopoverDirective.prototype.onClick = function (e) {
        e.preventDefault();
        this.openPopOver();
    };
    PopoverDirective.prototype.openPopOver = function () {
        var _this = this;
        if (this.componentRef) {
            this.closePopOver();
            return;
        }
        var componentFactory = this.factoryResolver.resolveComponentFactory(PopoverComponent);
        this.componentRef = this.viewContainerRef.createComponent(componentFactory);
        this.componentRef.instance.anchor = this.anchor || this.viewContainerRef.element;
        this.componentRef.instance.content = this.content;
        this.componentRef.instance.primaryPos = this.position;
        this.closeSub = this.componentRef.instance.close.subscribe(function () {
            _this.closePopOver();
        });
        this.viewContainerRef.element.nativeElement.classList.add('open');
    };
    PopoverDirective.prototype.closePopOver = function () {
        if (this.componentRef) {
            if (this.closeSub) {
                this.closeSub.unsubscribe();
            }
            this.viewContainerRef.element.nativeElement.classList.remove('open');
            var viewIndex = this.viewContainerRef.indexOf(this.componentRef.hostView);
            if (viewIndex !== -1) {
                this.viewContainerRef.remove(viewIndex);
            }
            this.componentRef = null;
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], PopoverDirective.prototype, "position", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], PopoverDirective.prototype, "anchor", void 0);
    tslib_1.__decorate([
        Input('rbPopover'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PopoverDirective.prototype, "template", null);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [MouseEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], PopoverDirective.prototype, "onClick", null);
    PopoverDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbPopover]',
            exportAs: 'rbPopover'
        }),
        tslib_1.__metadata("design:paramtypes", [ComponentFactoryResolver,
            ViewContainerRef])
    ], PopoverDirective);
    return PopoverDirective;
}());
export { PopoverDirective };
