// @inst-iot/bosch-angular-ui-components/esm5/molecules/tab-panel/tab-panel-item.directive.js
import * as tslib_1 from "tslib";
import { ContentChild, Directive, Input, TemplateRef } from '@angular/core';
import { TabPanelTitleDirective } from './tab-panel-title.directive';
var tabItemNum = 0;
var TabPanelItemDirective = /** @class */ (function () {
    function TabPanelItemDirective(templateRef) {
        this.templateRef = templateRef;
        this.rbTabPanelItemPosition = 'left';
        this.rbTabPanelItemType = 'tab';
        this.rbTabPanelItemAutoClose = true; // dropdown option
        this.routerLink = null;
        this.active = false;
        this.shown = 0;
        this.titleTpl = null;
        this.titleStr = null;
        this.id = 'tabItem' + (tabItemNum++);
    }
    Object.defineProperty(TabPanelItemDirective.prototype, "titleItem", {
        set: function (title) {
            if (title && !this.titleTpl) {
                this.titleTpl = title.templateRef;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelItemDirective.prototype, "rbTabPanelItemDetails", {
        set: function (details) {
            this.details = details;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelItemDirective.prototype, "rbTabPanelItemId", {
        set: function (id) {
            this.id = id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelItemDirective.prototype, "rbTabPanelItem", {
        set: function (title) {
            if (title instanceof TemplateRef) {
                this.titleTpl = title;
            }
            else {
                this.titleStr = title;
            }
        },
        enumerable: true,
        configurable: true
    });
    TabPanelItemDirective.prototype.show = function () {
        this.active = true;
        this.shown++;
    };
    TabPanelItemDirective.prototype.hide = function () {
        this.active = false;
    };
    Object.defineProperty(TabPanelItemDirective.prototype, "position", {
        get: function () {
            return this.rbTabPanelItemPosition;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelItemDirective.prototype, "type", {
        get: function () {
            return this.rbTabPanelItemType;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelItemDirective.prototype, "autoClose", {
        get: function () {
            return this.rbTabPanelItemAutoClose;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], TabPanelItemDirective.prototype, "rbTabPanelItemPosition", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], TabPanelItemDirective.prototype, "rbTabPanelItemType", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], TabPanelItemDirective.prototype, "rbTabPanelItemAutoClose", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], TabPanelItemDirective.prototype, "rbTabPanelItemQueryParams", void 0);
    tslib_1.__decorate([
        ContentChild(TabPanelTitleDirective, { static: false }),
        tslib_1.__metadata("design:type", TabPanelTitleDirective),
        tslib_1.__metadata("design:paramtypes", [TabPanelTitleDirective])
    ], TabPanelItemDirective.prototype, "titleItem", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], TabPanelItemDirective.prototype, "rbTabPanelItemDetails", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], TabPanelItemDirective.prototype, "rbTabPanelItemId", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], TabPanelItemDirective.prototype, "rbTabPanelItem", null);
    TabPanelItemDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbTabPanelItem]'
        }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], TabPanelItemDirective);
    return TabPanelItemDirective;
}());
export { TabPanelItemDirective };
