// @inst-iot/bosch-angular-ui-components/esm5/molecules/tab-panel/tab-panel-router-outlet.directive.js
import * as tslib_1 from "tslib";
import { ChildrenOutletContexts, RouterOutlet } from '@angular/router';
import { Attribute, Directive } from '@angular/core';
/**
 * Mocks the Router Outlet
 */
var TabPanelRouterOutletDirective = /** @class */ (function (_super) {
    tslib_1.__extends(TabPanelRouterOutletDirective, _super);
    function TabPanelRouterOutletDirective(_parentContexts, name) {
        var _this = _super.call(this, _parentContexts, null, null, name, null) || this;
        _this._parentContexts = _parentContexts;
        _this.__activatedRoute = null;
        _this._name = name;
        return _this;
    }
    TabPanelRouterOutletDirective.prototype.ngOnInit = function () {
        if (!this.__activated) {
            // If the outlet was not instantiated at the time the route got activated we need to populate
            // the outlet when it is initialized (ie inside a NgIf)
            var context = this._parentContexts.getContext(this._name);
            if (context && context.route) {
                if (context.attachRef) {
                    // `attachRef` is populated when there is an existing component to mount
                    this.attach(context.attachRef, context.route);
                }
                else {
                    // otherwise the component defined in the configuration is created
                    this.activateWith(context.route, context.resolver || null);
                }
            }
        }
    };
    Object.defineProperty(TabPanelRouterOutletDirective.prototype, "isActivated", {
        get: function () {
            return this.__activated;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelRouterOutletDirective.prototype, "component", {
        get: function () {
            if (!this.__activated) {
                throw new Error('Outlet is not activated');
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelRouterOutletDirective.prototype, "activatedRoute", {
        get: function () {
            if (!this.__activated) {
                throw new Error('Outlet is not activated');
            }
            return this.__activatedRoute;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPanelRouterOutletDirective.prototype, "activatedRouteData", {
        get: function () {
            if (this.__activatedRoute) {
                return this.__activatedRoute.snapshot.data;
            }
            return {};
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Called when the `RouteReuseStrategy` instructs to detach the subtree
     */
    TabPanelRouterOutletDirective.prototype.detach = function () {
        if (!this.__activated) {
            throw new Error('Outlet is not activated');
        }
        this.__activated = null;
        this.__activatedRoute = null;
        return null;
    };
    /**
     * Called when the `RouteReuseStrategy` instructs to re-attach a previously detached subtree
     */
    TabPanelRouterOutletDirective.prototype.attach = function (ref, activatedRoute) {
        this.__activated = true;
        this.__activatedRoute = activatedRoute;
        this.onActivate(activatedRoute);
    };
    TabPanelRouterOutletDirective.prototype.deactivate = function () {
        if (this.__activated) {
            this.__activated = false;
            this.__activatedRoute = null;
        }
    };
    TabPanelRouterOutletDirective.prototype.activateWith = function (activatedRoute, resolver) {
        console.log('activateWith', activatedRoute);
        if (this.isActivated) {
            throw new Error('Cannot activate an already activated outlet');
        }
        this.__activatedRoute = activatedRoute;
        this.__activated = true;
        this.onActivate(activatedRoute);
    };
    TabPanelRouterOutletDirective.prototype.onActivate = function (activatedRoute) {
    };
    TabPanelRouterOutletDirective = tslib_1.__decorate([
        Directive({
            // tslint:disable-next-line
            selector: 'rb-tab-panel-outlet'
        }),
        tslib_1.__param(1, Attribute('name')),
        tslib_1.__metadata("design:paramtypes", [ChildrenOutletContexts, String])
    ], TabPanelRouterOutletDirective);
    return TabPanelRouterOutletDirective;
}(RouterOutlet));
export { TabPanelRouterOutletDirective };
