// @inst-iot/bosch-angular-ui-components/esm5/organisms/full-header/full-header.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, ElementRef, HostListener, Optional, ViewChild } from '@angular/core';
import { ActionNavItemsDirective, MainNavItemsDirective, MetaNavItemsDirective, SubNavItemsDirective } from './navigation/navigation.directive';
import { LogoHeaderDirective, SubBrandHeaderDirective } from './full-header.directive';
import { NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
var FullHeaderComponent = /** @class */ (function () {
    function FullHeaderComponent(router) {
        this.router = router;
        this.open = false;
        this.opened = false;
        this.fullHeight = null;
    }
    FullHeaderComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (this.router) {
            this.router.events.pipe(filter(function (event) { return event instanceof NavigationEnd; })).subscribe(function (event) {
                _this.open = false;
            });
        }
    };
    FullHeaderComponent.prototype.toggle = function () {
        this.open = !this.open;
        if (this.open) {
            this.opened = true;
        }
    };
    FullHeaderComponent.prototype.onScroll = function () {
        var header = this.headerElement.nativeElement;
        var placeholder = this.headerPlaceholderElement.nativeElement;
        var height = header.offsetHeight;
        var isSticky = header.classList.contains('sticky');
        if (!isSticky) {
            this.fullHeight = height;
        }
        if (window.pageYOffset > this.fullHeight && !isSticky) {
            header.classList.add('sticky');
            placeholder.classList.add('sticky');
            placeholder.style.height = this.fullHeight + 'px';
        }
        else if (window.pageYOffset <= this.fullHeight && isSticky) {
            header.classList.remove('sticky');
            placeholder.classList.remove('sticky');
        }
    };
    tslib_1.__decorate([
        ViewChild('header', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], FullHeaderComponent.prototype, "headerElement", void 0);
    tslib_1.__decorate([
        ViewChild('placeholder', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], FullHeaderComponent.prototype, "headerPlaceholderElement", void 0);
    tslib_1.__decorate([
        ContentChild(MainNavItemsDirective, { static: false }),
        tslib_1.__metadata("design:type", MainNavItemsDirective)
    ], FullHeaderComponent.prototype, "mainNavItems", void 0);
    tslib_1.__decorate([
        ContentChild(SubNavItemsDirective, { static: false }),
        tslib_1.__metadata("design:type", SubNavItemsDirective)
    ], FullHeaderComponent.prototype, "subNavItems", void 0);
    tslib_1.__decorate([
        ContentChild(MetaNavItemsDirective, { static: false }),
        tslib_1.__metadata("design:type", MetaNavItemsDirective)
    ], FullHeaderComponent.prototype, "metaNavItems", void 0);
    tslib_1.__decorate([
        ContentChild(ActionNavItemsDirective, { static: false }),
        tslib_1.__metadata("design:type", ActionNavItemsDirective)
    ], FullHeaderComponent.prototype, "actionNavItems", void 0);
    tslib_1.__decorate([
        ContentChild(SubBrandHeaderDirective, { static: false }),
        tslib_1.__metadata("design:type", SubBrandHeaderDirective)
    ], FullHeaderComponent.prototype, "subBrand", void 0);
    tslib_1.__decorate([
        ContentChild(LogoHeaderDirective, { static: false }),
        tslib_1.__metadata("design:type", LogoHeaderDirective)
    ], FullHeaderComponent.prototype, "customLogo", void 0);
    tslib_1.__decorate([
        HostListener('window:scroll'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], FullHeaderComponent.prototype, "onScroll", null);
    FullHeaderComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-full-header',
            template: "<div class=\"header-placeholder\" #placeholder></div>\r\n<header class=\"header\" #header>\r\n  <div class=\"supergraphic bar\"></div>\r\n  <div class=\"header-content limit\">\r\n    <div class=\"logo\" *ngIf=\"customLogo\">\r\n      <ng-container *ngTemplateOutlet=\"customLogo.templateRef\"></ng-container>\r\n    </div>\r\n    <div class=\"logo\" *ngIf=\"!customLogo\">\r\n      <a href=\"https://www.bosch.com/\" target=\"_blank\">\r\n        <img src=\"assets/img/Bosch_EN_LV_RGB.png\" alt=\"Bosch - Invented for life\">\r\n      </a>\r\n    </div>\r\n    <div class=\"sub-brand\" *ngIf=\"subBrand\">\r\n      <div class=\"sub-brand-content\">\r\n        <ng-container *ngTemplateOutlet=\"subBrand.templateRef\"></ng-container>\r\n      </div>\r\n    </div>\r\n    <div class=\"small-nav\">\r\n      <button class=\"toggler\" [class.open]=\"open\" (click)=\"toggle()\">\r\n        <span class=\"bar\"></span>\r\n        <span class=\"bar\"></span>\r\n        <span class=\"bar\"></span>\r\n      </button>\r\n    </div>\r\n    <div class=\"main-nav\" [class.open]=\"open\" [class.opened]=\"opened\" *ngIf=\"mainNavItems\">\r\n      <rb-navigation type=\"main\" [itemsTemplate]=\"mainNavItems\"></rb-navigation>\r\n    </div>\r\n    <div class=\"action-nav\">\r\n      <rb-navigation type=\"action\" [itemsTemplate]=\"actionNavItems\"></rb-navigation>\r\n    </div>\r\n    <div class=\"sub-nav\" [class.open]=\"open\" [class.opened]=\"opened\" *ngIf=\"subNavItems\">\r\n      <rb-navigation type=\"sub\" [itemsTemplate]=\"subNavItems\"></rb-navigation>\r\n    </div>\r\n    <div class=\"meta-nav\" [class.open]=\"open\" [class.opened]=\"opened\" *ngIf=\"metaNavItems\">\r\n      <rb-navigation type=\"meta\" [itemsTemplate]=\"metaNavItems\"></rb-navigation>\r\n    </div>\r\n  </div>\r\n\r\n</header>\r\n"
        }),
        tslib_1.__param(0, Optional()),
        tslib_1.__metadata("design:paramtypes", [Router])
    ], FullHeaderComponent);
    return FullHeaderComponent;
}());
export { FullHeaderComponent };
