// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/form-validation-message.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, TemplateRef } from '@angular/core';
let FormValidationMessageDirective = class FormValidationMessageDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.onTouched = false;
    }
    set rbFormValidationMessage(name) {
        this.name = name;
    }
    set rbFormValidationMessageOnTouched(onTouched) {
        this.onTouched = onTouched;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], FormValidationMessageDirective.prototype, "rbFormValidationMessage", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], FormValidationMessageDirective.prototype, "rbFormValidationMessageOnTouched", null);
FormValidationMessageDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbFormValidationMessage]'
    }),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], FormValidationMessageDirective);
export { FormValidationMessageDirective };
