// @inst-iot/bosch-angular-ui-components/esm5/atoms/dropdown/dropdown.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, HostListener, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { getAbsolutePosition } from '../../utils/positionUtils';
var DropdownComponent = /** @class */ (function () {
    function DropdownComponent() {
        this.primaryPos = 'left';
        this.autoClose = true;
        this.hugContent = false;
        this.close = new EventEmitter();
        this.shown = false;
        this.updateContext({});
    }
    DropdownComponent.prototype.onAnyClick = function () {
        if (this.shown) {
            this.doClose('click');
        }
    };
    DropdownComponent.prototype.onClick = function (event) {
        event.stopPropagation();
        if (this.shown && this.autoClose) {
            this.doClose('click');
        }
    };
    DropdownComponent.prototype.onResize = function () {
        this.positionToAnchor();
    };
    DropdownComponent.prototype.ngOnInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.positionToAnchor();
        });
        this.positionToAnchor();
    };
    DropdownComponent.prototype.show = function () {
        var _this = this;
        setTimeout(function () {
            _this.shown = true;
        });
        this.popoverElementRef.nativeElement.classList.add('open');
        this.positionToAnchor();
    };
    DropdownComponent.prototype.hide = function () {
        this.shown = false;
        this.popoverElementRef.nativeElement.classList.remove('open');
    };
    DropdownComponent.prototype.ngAfterViewChecked = function () {
        this.positionToAnchor();
    };
    DropdownComponent.prototype.positionToAnchor = function () {
        if (!this.anchor || !this.popoverElementRef) {
            return;
        }
        var a = this.anchor.nativeElement;
        var pop = this.popoverElementRef.nativeElement;
        var anchorPos = getAbsolutePosition(a);
        var viewTop = window.pageYOffset;
        var viewBottom = window.innerHeight + window.pageYOffset;
        var isBottom = true;
        if (!this.hugContent) {
            pop.style.minWidth = a.offsetWidth + 'px';
        }
        var pos = {
            x: anchorPos.x,
            y: 0
        };
        if (this.primaryPos === 'center') {
            pos.x += a.offsetWidth / 2 - pop.offsetWidth / 2;
        }
        if (this.primaryPos === 'right') {
            pos.x += a.offsetWidth - pop.offsetWidth;
        }
        pos.y = anchorPos.y + a.offsetHeight;
        if (pos.y + pop.offsetHeight > viewBottom) {
            isBottom = false;
        }
        if (!isBottom) {
            pos.y = anchorPos.y - pop.offsetHeight;
            if (pos.y < 0) {
                isBottom = true;
            }
        }
        if (isBottom) {
            pos.y = anchorPos.y + a.offsetHeight;
        }
        else {
        }
        if (pos.x + pop.offsetWidth > window.innerWidth) {
            pos.x = window.innerWidth - pop.offsetWidth;
        }
        if (pos.x < 0) {
            pos.x = 0;
        }
        var popParentPos = getAbsolutePosition(pop.offsetParent);
        pos.y -= popParentPos.y;
        pos.x -= popParentPos.x;
        pop.style.top = pos.y + 'px';
        pop.style.left = pos.x + 'px';
    };
    DropdownComponent.prototype.doClose = function (reason) {
        this.close.next(reason);
    };
    DropdownComponent.prototype.updateContext = function (ctx) {
        this.context = tslib_1.__assign({}, ctx, { close: this.doClose.bind(this) });
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", ElementRef)
    ], DropdownComponent.prototype, "anchor", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", TemplateRef)
    ], DropdownComponent.prototype, "content", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], DropdownComponent.prototype, "primaryPos", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DropdownComponent.prototype, "autoClose", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DropdownComponent.prototype, "hugContent", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], DropdownComponent.prototype, "close", void 0);
    tslib_1.__decorate([
        ViewChild('dropdown', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], DropdownComponent.prototype, "popoverElementRef", void 0);
    tslib_1.__decorate([
        HostListener('window:click'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], DropdownComponent.prototype, "onAnyClick", null);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [MouseEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], DropdownComponent.prototype, "onClick", null);
    tslib_1.__decorate([
        HostListener('window:resize'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], DropdownComponent.prototype, "onResize", null);
    DropdownComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-dropdown',
            template: "<div class=\"dropdown\" #dropdown>\n  <div class=\"dropdown-content\">\n    <ng-container *ngTemplateOutlet=\"content; context: context\"></ng-container>\n  </div>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], DropdownComponent);
    return DropdownComponent;
}());
export { DropdownComponent };
