// @inst-iot/bosch-angular-ui-components/esm2015/molecules/breadcrumbs/breadcrumbs.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { BreadcrumbsService } from './breadcrumbs.service';
let BreadcrumbsComponent = class BreadcrumbsComponent {
    constructor(breadcrumbsService) {
        this.breadcrumbsService = breadcrumbsService;
        this.breadcrumbs = [];
        /**
         * Minimal amount of entries the breadcrumb must have to be displayed
         */
        this.minElements = 0;
        this.sub = null;
    }
    ngOnInit() {
        this.sub = this.breadcrumbsService.breadcrumbs.subscribe(list => {
            this.breadcrumbs = list;
        });
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], BreadcrumbsComponent.prototype, "minElements", void 0);
BreadcrumbsComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-breadcrumbs',
        template: "<ng-container *ngIf=\"breadcrumbs.length >= minElements\">\n  <div class=\"limit breadcrumbs-container\">\n    <ul class=\"breadcrumbs\">\n      <li *ngFor=\"let data of breadcrumbs\">\n        <a [routerLink]=\"data.urlSegments\">{{data.title}}</a>\n      </li>\n    </ul>\n    <div class=\"breadcrumbs-right\">\n      <ng-content></ng-content>\n    </div>\n  </div>\n  <hr class=\"no-margin\">\n</ng-container>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [BreadcrumbsService])
], BreadcrumbsComponent);
export { BreadcrumbsComponent };
