// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-date-fields/form-relative-time-input/relative-time-input.component.js
import * as tslib_1 from "tslib";
var RelativeTimeInputComponent_1;
import { Component, forwardRef, Inject, InjectionToken, Input, Optional, TemplateRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { defaultTimeInputOptions, getRelativeInfo } from '../date-range-picker/date-range-picker.model';
export const TIME_INPUT_OPTIONS = new InjectionToken('formRelativeInputTimeOptions');
let RelativeTimeInputComponent = RelativeTimeInputComponent_1 = class RelativeTimeInputComponent {
    constructor(options) {
        this.options = options;
        this.id = 'input.' + Math.random();
        this.onChange = emptyFunction;
        this.onTouched = emptyFunction;
        this.disabled = false;
        this.operator = '-';
        this.count = '0';
        this.unit = 'ms';
        this.displayText = '';
        if (!options) {
            this.options = defaultTimeInputOptions;
        }
    }
    ngOnInit() {
        this.units = defaultTimeInputOptions.units;
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    writeValue(obj) {
        const info = getRelativeInfo(obj, this.units);
        this.count = info.count.toString();
        this.operator = info.operator;
        this.unit = info.unit;
        this.displayText = info.displayText;
    }
    getValue() {
        const count = parseInt(this.count, 10) * this.getDivisor();
        if (isNaN(count)) {
            return 0;
        }
        else {
            return this.operator === '-' ? count * -1 : count;
        }
    }
    getDivisor() {
        return this.units.find(u => u.unit === this.unit).value;
    }
    notifyChange() {
        const value = this.getValue();
        this.onChange(value);
    }
    updateUnit(unit) {
        const oldDivisor = this.getDivisor();
        this.unit = unit;
        const newDivisor = this.getDivisor();
        if (newDivisor > oldDivisor) {
            const baseValue = parseInt(this.count, 10) * oldDivisor;
            this.count = Math.ceil(baseValue / newDivisor).toString();
        }
        this.notifyChange();
    }
    isLabelTemplate() {
        return this.label instanceof TemplateRef;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RelativeTimeInputComponent.prototype, "label", void 0);
RelativeTimeInputComponent = RelativeTimeInputComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'rb-form-relative-time-input',
        template: "<div class=\"rb-custom-input\">\r\n  <div class=\"input-wrapper\" [class.no-label]=\"!label\">\r\n    <div class=\"relative-time-form\">\r\n      <rb-form-select [(ngModel)]=\"operator\" (ngModelChange)=\"notifyChange()\" name=\"operator\" [disabled]=\"disabled\">\r\n        <option value=\"-\">-</option>\r\n        <option value=\"+\">+</option>\r\n      </rb-form-select>\r\n      <rb-form-input class=\"relative-time-form-number\" type=\"number\"\r\n                     [(ngModel)]=\"count\" (ngModelChange)=\"notifyChange()\" name=\"count\"\r\n                     pattern=\"^\\d+$\" required\r\n                     [disabled]=\"disabled\"></rb-form-input>\r\n      <rb-form-select class=\"relative-time-form-unit\" [disabled]=\"disabled\"\r\n                      [ngModel]=\"unit\"\r\n                      (ngModelChange)=\"updateUnit($event)\"\r\n                      name=\"unit\">\r\n        <option *ngFor=\"let u of units\" [value]=\"u.unit\">{{u.label}}</option>\r\n      </rb-form-select>\r\n    </div>\r\n\r\n    <label class=\"label\" [for]=\"id + 'input'\">\r\n      {{!isLabelTemplate()?label:''}}\r\n      <ng-container *ngIf=\"isLabelTemplate()\">\r\n        <ng-container *ngTemplateOutlet=\"label\"></ng-container>\r\n      </ng-container>\r\n    </label>\r\n\r\n    <span class=\"input-background\"></span>\r\n\r\n  </div>\r\n</div>\r\n",
        providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => RelativeTimeInputComponent_1), multi: true }]
    }),
    tslib_1.__param(0, Inject(TIME_INPUT_OPTIONS)), tslib_1.__param(0, Optional()),
    tslib_1.__metadata("design:paramtypes", [Object])
], RelativeTimeInputComponent);
export { RelativeTimeInputComponent };
export function emptyFunction(_) {
}
