// @inst-iot/bosch-angular-ui-components/esm2015/atoms/form-fields/form-custom-select/form-select-option/form-select-option.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Inject, Input, ViewChild } from '@angular/core';
import { FORM_SELECT_SELECTOR } from '../form-custom-select.model';
import { Subject } from 'rxjs';
let FormSelectOptionComponent = class FormSelectOptionComponent {
    constructor(selectSubject, elementRef) {
        this.selectSubject = selectSubject;
        this.elementRef = elementRef;
        this.active = false;
    }
    ngOnInit() {
        this.sub = this.selectSubject.subscribe(action => {
            this.active = action.value === this.value;
        });
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
    select() {
        this.selectSubject.next({ type: 'update', value: this.value });
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], FormSelectOptionComponent.prototype, "value", void 0);
tslib_1.__decorate([
    ViewChild('innerLink', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], FormSelectOptionComponent.prototype, "content", void 0);
FormSelectOptionComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-form-select-option',
        template: "<a href=\"javascript:\" class=\"rb-dropdown-item\" [class.active]=\"active\" (click)=\"select()\" #innerLink>\r\n  <ng-content></ng-content>\r\n</a>\r\n"
    }),
    tslib_1.__param(0, Inject(FORM_SELECT_SELECTOR)),
    tslib_1.__metadata("design:paramtypes", [Subject,
        ElementRef])
], FormSelectOptionComponent);
export { FormSelectOptionComponent };
