// @inst-iot/bosch-angular-ui-components/esm2015/atoms/modal/dialog/dialog.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { ModalService } from '../modal.service';
let DialogComponent = class DialogComponent {
    constructor(modalService) {
        this.modalService = modalService;
        this.showButtons = true;
    }
    ngOnInit() {
    }
    close(reason) {
        this.modalService.close(reason);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], DialogComponent.prototype, "dialogTitle", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], DialogComponent.prototype, "okBtnLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], DialogComponent.prototype, "cancelBtnLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], DialogComponent.prototype, "showButtons", void 0);
DialogComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-dialog',
        template: "<h2 class=\"dialog-title\">{{dialogTitle}}</h2>\n<ng-content></ng-content>\n<div class=\"dialog-buttons\" *ngIf=\"showButtons\">\n  <button class=\"rb-btn\" (click)=\"close()\">{{cancelBtnLabel ? cancelBtnLabel : 'Cancel'}}</button>&nbsp;\n  <button class=\"rb-btn rb-primary\" (click)=\"close(true)\">{{okBtnLabel ? okBtnLabel : 'OK'}}</button>\n</div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ModalService])
], DialogComponent);
export { DialogComponent };
