// @inst-iot/bosch-angular-ui-components/esm5/atoms/form-fields/form-file/form-file.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, ElementRef, EventEmitter, forwardRef, Input, Output, QueryList, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { FormValidationMessageDirective } from '../form-validation-message.directive';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { emptyFunction } from '../forms-util';
var FormFileComponent = /** @class */ (function () {
    function FormFileComponent(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.label = null;
        this.placeholder = '';
        this.filesChange = new EventEmitter();
        this.id = 'file.' + Math.random();
        this.onChange = emptyFunction;
        this.onTouched = emptyFunction;
        this.filePreview = '';
    }
    FormFileComponent_1 = FormFileComponent;
    FormFileComponent.prototype.checkSelectedFiles = function () {
        var fileNames = [];
        var files = [];
        for (var i = 0; i < this.file.nativeElement.files.length; i++) {
            var file = this.file.nativeElement.files[i];
            files.push(file);
            if (file.name) {
                fileNames.push(file.name);
            }
        }
        if (files.length === 0) {
            this.renderer.removeClass(this.elementRef.nativeElement, 'not-empty');
        }
        else {
            this.renderer.addClass(this.elementRef.nativeElement, 'not-empty');
        }
        this.filePreview = fileNames.join(', ');
        if (this.filePreview.length > 100) {
            this.filePreview = this.filePreview.substr(0, 97) + '... (' + fileNames.length + ')';
        }
        console.log('files', files);
        this.onChange(files);
    };
    FormFileComponent.prototype.isLabelTemplate = function () {
        return this.label instanceof TemplateRef;
    };
    FormFileComponent.prototype.writeValue = function (value) {
        // console.warn('unsupported write to rb-form-file', value);
    };
    FormFileComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    FormFileComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    FormFileComponent.prototype.setDisabledState = function (isDisabled) {
        this.renderer.setProperty(this.file.nativeElement, 'disabled', isDisabled);
    };
    var FormFileComponent_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormFileComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], FormFileComponent.prototype, "name", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormFileComponent.prototype, "placeholder", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], FormFileComponent.prototype, "filesChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], FormFileComponent.prototype, "accept", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], FormFileComponent.prototype, "multiple", void 0);
    tslib_1.__decorate([
        ViewChild('file', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], FormFileComponent.prototype, "file", void 0);
    tslib_1.__decorate([
        ContentChildren(FormValidationMessageDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], FormFileComponent.prototype, "messages", void 0);
    FormFileComponent = FormFileComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'rb-form-file',
            template: "<div class=\"input-wrapper\" [class.no-label]=\"!label\">\r\n\r\n  <input type=\"file\"\r\n         [id]=\"id + 'input'\"\r\n         [placeholder]=\"placeholder\"\r\n         [multiple]=\"multiple !== undefined\"\r\n         [accept]=\"accept\"\r\n         class=\"file-input with-icon\"\r\n         #file\r\n         (change)=\"checkSelectedFiles()\"\r\n         (blur)=\"onTouched()\">\r\n  <span class=\"input-icon no-pointer\">\r\n    <span class=\"rb-ic rb-ic-upload\"></span>\r\n  </span>\r\n\r\n  <label class=\"rb-btn rb-select-btn input\" [for]=\"id + 'input'\">{{filePreview}}</label>\r\n\r\n  <label class=\"label\" [for]=\"id + 'input'\">\r\n    {{!isLabelTemplate()?label:''}}\r\n    <ng-container *ngIf=\"isLabelTemplate()\">\r\n      <ng-container *ngTemplateOutlet=\"label\"></ng-container>\r\n    </ng-container>\r\n  </label>\r\n  <span class=\"input-background\"></span>\r\n</div>\r\n<rb-form-errors [messages]=\"messages\"></rb-form-errors>\r\n",
            providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(function () { return FormFileComponent_1; }), multi: true }]
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2, ElementRef])
    ], FormFileComponent);
    return FormFileComponent;
}());
export { FormFileComponent };
