// @inst-iot/bosch-angular-ui-components/esm5/atoms/modal/modal.directive.js
import * as tslib_1 from "tslib";
import { ApplicationRef, ComponentFactoryResolver, Directive, HostListener, Injector, Input, Optional } from '@angular/core';
import { ModalService } from './modal.service';
var ModalDirective = /** @class */ (function () {
    function ModalDirective(modalService, factoryResolver, injector, appRef) {
        this.modalService = modalService;
        this.showClose = true;
        this.position = 'middle';
        this.backdropClose = true;
        this.stacked = false;
        if (!this.modalService) {
            this.modalService = new ModalService(factoryResolver, injector, appRef);
        }
    }
    ModalDirective.prototype.click = function (event) {
        event.preventDefault();
        this.modalService.open(this.content, {
            showClose: this.showClose,
            position: this.position,
            maxWidth: this.maxWidth,
            backdropClose: this.backdropClose,
            stacked: this.stacked
        });
    };
    Object.defineProperty(ModalDirective.prototype, "rbModal", {
        set: function (content) {
            this.content = content;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ModalDirective.prototype, "showClose", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], ModalDirective.prototype, "position", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], ModalDirective.prototype, "maxWidth", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ModalDirective.prototype, "backdropClose", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ModalDirective.prototype, "stacked", void 0);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [MouseEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], ModalDirective.prototype, "click", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], ModalDirective.prototype, "rbModal", null);
    ModalDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbModal]'
        }),
        tslib_1.__param(0, Optional()),
        tslib_1.__metadata("design:paramtypes", [ModalService,
            ComponentFactoryResolver,
            Injector,
            ApplicationRef])
    ], ModalDirective);
    return ModalDirective;
}());
export { ModalDirective };
