// @inst-iot/bosch-angular-ui-components/esm5/atoms/modal/modal.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, ViewChild, ViewContainerRef } from '@angular/core';
var ModalComponent = /** @class */ (function () {
    function ModalComponent(elementRef, viewContainerRef) {
        this.elementRef = elementRef;
        this.viewContainerRef = viewContainerRef;
        this.options = {
            position: 'middle',
            showClose: true,
            backdropClose: true
        };
    }
    ModalComponent.prototype.ngOnInit = function () {
        if (this.options.maxWidth) {
            this.elementRef.nativeElement.style.maxWidth = this.options.maxWidth;
        }
        if (this.options.position) {
            this.elementRef.nativeElement.classList.add(this.options.position);
        }
        if (this.contentComponentRef && this.box) {
            this.viewContainerRef.insert(this.contentComponentRef.hostView);
        }
    };
    ModalComponent.prototype.ngAfterViewInit = function () {
        if (this.contentComponentRef && this.box) {
            this.box.nativeElement.appendChild(this.contentComponentRef.location.nativeElement);
        }
    };
    tslib_1.__decorate([
        ViewChild('box', { static: true }),
        tslib_1.__metadata("design:type", ElementRef)
    ], ModalComponent.prototype, "box", void 0);
    ModalComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-modal',
            template: "<div class=\"modal-box\" role=\"dialog\" #box>\n  <button class=\"close-btn rb-btn rb-passive-link\" (click)=\"close()\" *ngIf=\"options.showClose\"><rb-icon name=\"close\"></rb-icon></button>\n\n  <ng-container *ngIf=\"contentTpl\">\n    <ng-container *ngTemplateOutlet=\"contentTpl; context: {close: close}\"></ng-container>\n  </ng-container>\n  <ng-container *ngIf=\"contentComponent\">\n    <ng-container *ngComponentOutlet=\"contentComponent\"></ng-container>\n  </ng-container>\n  {{contentText ? contentText : ''}}\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, ViewContainerRef])
    ], ModalComponent);
    return ModalComponent;
}());
export { ModalComponent };
