// @inst-iot/bosch-angular-ui-components/esm2015/atoms/modal/alert/alert.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { ModalService } from '../modal.service';
let AlertComponent = class AlertComponent {
    constructor(modalService) {
        this.modalService = modalService;
        this.type = 'success';
    }
    ngOnInit() {
    }
    close(reason) {
        this.modalService.close(reason);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], AlertComponent.prototype, "alertTitle", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], AlertComponent.prototype, "okBtnLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], AlertComponent.prototype, "cancelBtnLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], AlertComponent.prototype, "type", void 0);
AlertComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-alert',
        template: "<rb-icon [name]=\"type + '-filled'\"></rb-icon>\n<h2 class=\"dialog-title\">{{alertTitle}}</h2>\n<div class=\"dialog-text\">\n  <ng-content></ng-content>\n</div>\n<div class=\"dialog-buttons\">\n  <button class=\"rb-btn\" (click)=\"close(true)\" *ngIf=\"okBtnLabel\">{{okBtnLabel}}</button>\n  <button class=\"rb-btn\" (click)=\"close()\" *ngIf=\"cancelBtnLabel\">{{cancelBtnLabel}}</button>\n</div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [ModalService])
], AlertComponent);
export { AlertComponent };
