// @inst-iot/bosch-angular-ui-components/esm5/molecules/loading/error-format.model.js
import { InjectionToken } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
export var ERROR_FORMATTER = new InjectionToken('rbErrorFormatter');
var DetailedError = /** @class */ (function () {
    function DetailedError(message, details) {
        this.message = message;
        this.details = details;
        this.name = 'DetailedError';
    }
    DetailedError.prototype.toString = function () {
        return this.message;
    };
    return DetailedError;
}());
export { DetailedError };
var DefaultErrorFormatter = /** @class */ (function () {
    function DefaultErrorFormatter() {
    }
    DefaultErrorFormatter.prototype.getDetails = function (error) {
        if (error instanceof DetailedError) {
            if (typeof (error.details) === 'object') {
                return JSON.stringify(error.details, null, 2);
            }
            return error.details;
        }
        if (error instanceof HttpErrorResponse) {
            if (error.error) {
                if (typeof (error.error) === 'object') {
                    return JSON.stringify(error.error, null, 2);
                }
                return String(error.error);
            }
        }
        return null;
    };
    DefaultErrorFormatter.prototype.transform = function (error) {
        if (error instanceof HttpErrorResponse) {
            var text = 'Error: ';
            if (error.message) {
                text += error.message;
            }
            return text;
        }
        if (error instanceof Error || error instanceof DetailedError) {
            return String(error);
        }
        return null;
    };
    return DefaultErrorFormatter;
}());
export { DefaultErrorFormatter };
export var DEFAULT_ERROR_FORMATTER = new DefaultErrorFormatter();
