// @inst-iot/bosch-angular-ui-components/esm5/molecules/loading/loading-utils.js
import { Subject } from 'rxjs';
import { takeUntil, tap } from 'rxjs/operators';
var LoadingEntity = /** @class */ (function () {
    function LoadingEntity(loadingStateAfterFirstSuccess) {
        if (loadingStateAfterFirstSuccess === void 0) { loadingStateAfterFirstSuccess = true; }
        this.loadingStateAfterFirstSuccess = loadingStateAfterFirstSuccess;
        this.runs = 0;
        this.loading = false;
        this.error = null;
        this.result = null;
        this.stopSubject = new Subject();
    }
    /**
     * Tracks the status of an observable
     */
    LoadingEntity.prototype.run = function (observable) {
        var _this = this;
        this.runs++;
        if (this.loadingStateAfterFirstSuccess && this.result !== null || this.result === null || this.error) {
            this.loading = true;
            this.result = null;
        }
        this.complete();
        this.error = null;
        return observable.pipe(tap(function (result) {
            _this.loading = false;
            _this.result = result;
        }, function (error) {
            _this.loading = false;
            _this.error = error;
        }, function () {
            _this.loading = false;
        }), takeUntil(this.stopSubject));
    };
    /**
     * This can complete and unsubscribe all observers
     */
    LoadingEntity.prototype.complete = function () {
        this.stopSubject.next();
    };
    LoadingEntity.prototype.reset = function () {
        this.complete();
        this.error = null;
        this.loading = false;
        this.result = null;
    };
    return LoadingEntity;
}());
export { LoadingEntity };
