// @inst-iot/bosch-angular-ui-components/esm5/molecules/loading/loading-link.directive.js
import * as tslib_1 from "tslib";
import { ContentChildren, Directive, ElementRef, Input, QueryList, Renderer2 } from '@angular/core';
import { NavigationEnd, NavigationStart, Router, RouterLinkWithHref } from '@angular/router';
var LoadingLinkDirective = /** @class */ (function () {
    function LoadingLinkDirective(router, element, renderer) {
        var _this = this;
        this.router = router;
        this.element = element;
        this.renderer = renderer;
        this.loading = false;
        this.className = 'loading';
        this.subscription = router.events.subscribe(function (s) {
            if (s instanceof NavigationStart && _this.isRelevantUrl(s.url)) {
                _this.loading = true;
                _this.update();
            }
            if (s instanceof NavigationEnd) {
                _this.loading = false;
                _this.update();
            }
        });
    }
    Object.defineProperty(LoadingLinkDirective.prototype, "rbLoadingLink", {
        set: function (className) {
            if (className) {
                this.className = className;
            }
        },
        enumerable: true,
        configurable: true
    });
    LoadingLinkDirective.prototype.ngAfterContentInit = function () {
    };
    LoadingLinkDirective.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    LoadingLinkDirective.prototype.update = function () {
        if (!this.element || !this.element.nativeElement) {
            return;
        }
        if (this.loading) {
            this.renderer.addClass(this.element.nativeElement, this.className);
        }
        else {
            this.renderer.removeClass(this.element.nativeElement, this.className);
        }
    };
    LoadingLinkDirective.prototype.isRelevantUrl = function (url) {
        return this.linksWithHrefs.some(function (link) { return url === link.urlTree.toString(); });
    };
    tslib_1.__decorate([
        ContentChildren(RouterLinkWithHref, { descendants: true }),
        tslib_1.__metadata("design:type", QueryList)
    ], LoadingLinkDirective.prototype, "linksWithHrefs", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LoadingLinkDirective.prototype, "rbLoadingLink", null);
    LoadingLinkDirective = tslib_1.__decorate([
        Directive({
            selector: '[rbLoadingLink]'
        }),
        tslib_1.__metadata("design:paramtypes", [Router, ElementRef, Renderer2])
    ], LoadingLinkDirective);
    return LoadingLinkDirective;
}());
export { LoadingLinkDirective };
