// @inst-iot/bosch-angular-ui-components/esm2015/atoms/details-toggle/details-toggle.directive.js
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, HostListener, Input, Output, ViewContainerRef } from '@angular/core';
/**
 * Can be toggled to provide a state
 * Usage with: rbDetailsToggle #triggerDetails="rbDetailsToggle"
 * Then check triggerDetails.open for the current state
 */
let DetailsToggleDirective = class DetailsToggleDirective {
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this.isOpenChange = new EventEmitter();
        this.open = false;
    }
    set isOpen(state) {
        if (state) {
            this.openDetails();
        }
        else {
            this.closeDetails();
        }
    }
    onClick(e) {
        e.preventDefault();
        if (this.open) {
            this.closeDetails();
            this.isOpenChange.next(this.open);
        }
        else {
            this.openDetails();
            this.isOpenChange.next(this.open);
        }
    }
    openDetails() {
        this.open = true;
        this.viewContainerRef.element.nativeElement.classList.add('open');
    }
    closeDetails() {
        this.open = false;
        this.viewContainerRef.element.nativeElement.classList.remove('open');
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], DetailsToggleDirective.prototype, "isOpenChange", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], DetailsToggleDirective.prototype, "isOpen", null);
tslib_1.__decorate([
    HostListener('click', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [MouseEvent]),
    tslib_1.__metadata("design:returntype", void 0)
], DetailsToggleDirective.prototype, "onClick", null);
DetailsToggleDirective = tslib_1.__decorate([
    Directive({
        selector: '[rbDetailsToggle]',
        exportAs: 'rbDetailsToggle'
    }),
    tslib_1.__metadata("design:paramtypes", [ViewContainerRef])
], DetailsToggleDirective);
export { DetailsToggleDirective };
