// @inst-iot/bosch-angular-ui-components/esm5/atoms/callout/callout.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var CalloutComponent = /** @class */ (function () {
    function CalloutComponent() {
        this.type = 'error';
        this.icon = ['rb-ic', 'rb-ic-alert-error-filled'];
        this.type2icon = {
            error: 'rb-ic-alert-error-filled',
            success: 'rb-ic-alert-success-filled',
            warning: 'rb-ic-alert-warning-filled',
            info: 'rb-ic-info',
        };
    }
    CalloutComponent.prototype.ngOnChanges = function (changes) {
        if (changes.type) {
            this.icon = ['rb-ic', this.type2icon[changes.type.currentValue]];
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], CalloutComponent.prototype, "type", void 0);
    CalloutComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-callout',
            template: "<div [class]=\"'rb-callout rb-callout-'+type\">\r\n  <span [ngClass]=\"icon\"></span>\r\n  <ng-content></ng-content>\r\n</div>\r\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], CalloutComponent);
    return CalloutComponent;
}());
export { CalloutComponent };
