// @inst-iot/bosch-angular-ui-components/esm5/molecules/language-selector/language-selector.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var LanguageSelectorComponent = /** @class */ (function () {
    function LanguageSelectorComponent() {
        this.languageChange = new EventEmitter();
    }
    LanguageSelectorComponent.prototype.ngOnInit = function () {
    };
    LanguageSelectorComponent.prototype.selectLanguage = function (l) {
        this.language = l.id;
        this.languageChange.next(this.language);
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], LanguageSelectorComponent.prototype, "languages", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LanguageSelectorComponent.prototype, "language", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LanguageSelectorComponent.prototype, "languageChange", void 0);
    LanguageSelectorComponent = tslib_1.__decorate([
        Component({
            selector: 'rb-language-selector',
            template: "<a href=\"javascript:\" [rbDropdown]=\"items\" [position]=\"'right'\" class=\"language-selector-link language rb-dropdown-toggle\">\n  {{language}}\n</a>\n<ng-template #items>\n  <ul class=\"link-list\">\n    <li *ngFor=\"let l of languages\">\n      <a href=\"javascript:\" (click)=\"selectLanguage(l)\" [class.active]=\"l.id === language\">\n        <span class=\"language\">{{l.id}}</span> | {{l.name}}\n      </a>\n    </li>\n  </ul>\n</ng-template>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LanguageSelectorComponent);
    return LanguageSelectorComponent;
}());
export { LanguageSelectorComponent };
