// @inst-iot/bosch-angular-ui-components/esm2015/organisms/footer-nav/footer-nav.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, HostListener, Renderer2, ViewChild } from '@angular/core';
let FooterNavComponent = class FooterNavComponent {
    constructor(renderer) {
        this.renderer = renderer;
    }
    ngOnInit() {
        this.checkSticky();
    }
    ngDoCheck() {
        this.checkSticky();
    }
    onResize() {
        this.checkSticky();
    }
    checkSticky() {
        const footer = this.footerEl.nativeElement;
        const footerHeight = footer.offsetHeight;
        const windowHeight = window.innerHeight;
        const bottomBorder = footer.offsetTop + footer.offsetHeight;
        const realContentHeight = document.body.firstElementChild.offsetHeight;
        const sticky = footer.classList.contains('sticky');
        if (!sticky && bottomBorder < windowHeight) {
            this.renderer.addClass(footer, 'sticky');
            this.renderer.setStyle(this.placeholderEl.nativeElement, 'height', footerHeight + 'px');
        }
        else if (sticky && realContentHeight >= windowHeight) {
            this.renderer.removeClass(footer, 'sticky');
        }
    }
};
tslib_1.__decorate([
    ViewChild('footer', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], FooterNavComponent.prototype, "footerEl", void 0);
tslib_1.__decorate([
    ViewChild('placeholder', { static: true }),
    tslib_1.__metadata("design:type", ElementRef)
], FooterNavComponent.prototype, "placeholderEl", void 0);
tslib_1.__decorate([
    HostListener('window:resize'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], FooterNavComponent.prototype, "onResize", null);
FooterNavComponent = tslib_1.__decorate([
    Component({
        selector: 'rb-footer-nav',
        template: "<footer #footer>\n  <ng-content></ng-content>\n  <div class=\"supergraphic bar\"></div>\n</footer>\n<div class=\"footer-placeholder\" #placeholder></div>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [Renderer2])
], FooterNavComponent);
export { FooterNavComponent };
